*** /dev/null	Thu Sep 22 05:30:40 2016
--- test/configs/odroid-xu4_defconfig	Thu Sep 22 06:46:01 2016
***************
*** 0 ****
--- 1,682 ----
+ #
+ # Automatically generated file; DO NOT EDIT.
+ # U-Boot 2016.05 Configuration
+ #
+ CONFIG_CREATE_ARCH_SYMLINK=y
+ CONFIG_HAVE_GENERIC_BOARD=y
+ CONFIG_SYS_GENERIC_BOARD=y
+ # CONFIG_ARC is not set
+ CONFIG_ARM=y
+ # CONFIG_AVR32 is not set
+ # CONFIG_BLACKFIN is not set
+ # CONFIG_M68K is not set
+ # CONFIG_MICROBLAZE is not set
+ # CONFIG_MIPS is not set
+ # CONFIG_NDS32 is not set
+ # CONFIG_NIOS2 is not set
+ # CONFIG_OPENRISC is not set
+ # CONFIG_PPC is not set
+ # CONFIG_SANDBOX is not set
+ # CONFIG_SH is not set
+ # CONFIG_SPARC is not set
+ # CONFIG_X86 is not set
+ CONFIG_SYS_ARCH="arm"
+ CONFIG_SYS_CPU="armv7"
+ CONFIG_SYS_SOC="exynos"
+ CONFIG_SYS_VENDOR="samsung"
+ CONFIG_SYS_BOARD="smdk5420"
+ CONFIG_SYS_CONFIG_NAME="odroid_xu3"
+ 
+ #
+ # ARM architecture
+ #
+ CONFIG_HAS_VBAR=y
+ CONFIG_HAS_THUMB2=y
+ CONFIG_CPU_V7=y
+ # CONFIG_SEMIHOSTING is not set
+ # CONFIG_SYS_L2CACHE_OFF is not set
+ # CONFIG_ARCH_AT91 is not set
+ # CONFIG_TARGET_EDB93XX is not set
+ # CONFIG_TARGET_VCMA9 is not set
+ # CONFIG_TARGET_SMDK2410 is not set
+ # CONFIG_TARGET_ASPENITE is not set
+ # CONFIG_TARGET_GPLUGD is not set
+ # CONFIG_ARCH_DAVINCI is not set
+ # CONFIG_KIRKWOOD is not set
+ # CONFIG_ARCH_MVEBU is not set
+ # CONFIG_TARGET_DEVKIT3250 is not set
+ # CONFIG_TARGET_WORK_92105 is not set
+ # CONFIG_TARGET_MX25PDK is not set
+ # CONFIG_TARGET_ZMX25 is not set
+ # CONFIG_TARGET_APF27 is not set
+ # CONFIG_TARGET_APX4DEVKIT is not set
+ # CONFIG_TARGET_XFI3 is not set
+ # CONFIG_TARGET_M28EVK is not set
+ # CONFIG_TARGET_MX23EVK is not set
+ # CONFIG_TARGET_MX28EVK is not set
+ # CONFIG_TARGET_MX23_OLINUXINO is not set
+ # CONFIG_TARGET_BG0900 is not set
+ # CONFIG_TARGET_SANSA_FUZE_PLUS is not set
+ # CONFIG_TARGET_SC_SPS_1 is not set
+ # CONFIG_ORION5X is not set
+ # CONFIG_TARGET_SPEAR300 is not set
+ # CONFIG_TARGET_SPEAR310 is not set
+ # CONFIG_TARGET_SPEAR320 is not set
+ # CONFIG_TARGET_SPEAR600 is not set
+ # CONFIG_TARGET_STV0991 is not set
+ # CONFIG_TARGET_X600 is not set
+ # CONFIG_TARGET_IMX31_PHYCORE is not set
+ # CONFIG_TARGET_MX31ADS is not set
+ # CONFIG_TARGET_MX31PDK is not set
+ # CONFIG_TARGET_WOODBURN is not set
+ # CONFIG_TARGET_WOODBURN_SD is not set
+ # CONFIG_TARGET_FLEA3 is not set
+ # CONFIG_TARGET_MX35PDK is not set
+ # CONFIG_ARCH_BCM283X is not set
+ # CONFIG_TARGET_VEXPRESS_CA15_TC2 is not set
+ # CONFIG_TARGET_VEXPRESS_CA5X2 is not set
+ # CONFIG_TARGET_VEXPRESS_CA9X4 is not set
+ # CONFIG_TARGET_KWB is not set
+ # CONFIG_TARGET_TSERIES is not set
+ # CONFIG_TARGET_CM_T335 is not set
+ # CONFIG_TARGET_PEPPER is not set
+ # CONFIG_TARGET_AM335X_IGEP0033 is not set
+ # CONFIG_TARGET_PCM051 is not set
+ # CONFIG_TARGET_DRACO is not set
+ # CONFIG_TARGET_THUBAN is not set
+ # CONFIG_TARGET_RASTABAN is not set
+ # CONFIG_TARGET_PXM2 is not set
+ # CONFIG_TARGET_RUT is not set
+ # CONFIG_TARGET_PENGWYN is not set
+ # CONFIG_TARGET_AM335X_BALTOS is not set
+ # CONFIG_TARGET_AM335X_EVM is not set
+ # CONFIG_TARGET_AM335X_SL50 is not set
+ # CONFIG_TARGET_AM43XX_EVM is not set
+ # CONFIG_TARGET_BAV335X is not set
+ # CONFIG_TARGET_TI814X_EVM is not set
+ # CONFIG_TARGET_TI816X_EVM is not set
+ # CONFIG_TARGET_BCM28155_AP is not set
+ # CONFIG_TARGET_BCMCYGNUS is not set
+ # CONFIG_TARGET_BCMNSP is not set
+ CONFIG_ARCH_EXYNOS=y
+ # CONFIG_ARCH_S5PC1XX is not set
+ # CONFIG_ARCH_HIGHBANK is not set
+ # CONFIG_ARCH_INTEGRATOR is not set
+ # CONFIG_ARCH_KEYSTONE is not set
+ # CONFIG_ARCH_MX7 is not set
+ # CONFIG_ARCH_MX6 is not set
+ # CONFIG_ARCH_MX5 is not set
+ # CONFIG_TARGET_M53EVK is not set
+ # CONFIG_TARGET_MX51EVK is not set
+ # CONFIG_TARGET_MX53ARD is not set
+ # CONFIG_TARGET_MX53EVK is not set
+ # CONFIG_TARGET_MX53LOCO is not set
+ # CONFIG_TARGET_MX53SMD is not set
+ # CONFIG_OMAP34XX is not set
+ # CONFIG_OMAP44XX is not set
+ # CONFIG_OMAP54XX is not set
+ # CONFIG_RMOBILE is not set
+ # CONFIG_ARCH_SNAPDRAGON is not set
+ # CONFIG_ARCH_SOCFPGA is not set
+ # CONFIG_TARGET_CM_T43 is not set
+ # CONFIG_ARCH_SUNXI is not set
+ # CONFIG_TARGET_TS4800 is not set
+ # CONFIG_TARGET_VF610TWR is not set
+ # CONFIG_TARGET_COLIBRI_VF is not set
+ # CONFIG_TARGET_PCM052 is not set
+ # CONFIG_ARCH_ZYNQ is not set
+ # CONFIG_ARCH_ZYNQMP is not set
+ # CONFIG_TEGRA is not set
+ # CONFIG_TARGET_VEXPRESS64_AEMV8A is not set
+ # CONFIG_TARGET_VEXPRESS64_BASE_FVP is not set
+ # CONFIG_TARGET_VEXPRESS64_BASE_FVP_DRAM is not set
+ # CONFIG_TARGET_VEXPRESS64_JUNO is not set
+ # CONFIG_TARGET_LS2080A_EMU is not set
+ # CONFIG_TARGET_LS2080A_SIMU is not set
+ # CONFIG_TARGET_LS2080AQDS is not set
+ # CONFIG_TARGET_LS2080ARDB is not set
+ # CONFIG_TARGET_HIKEY is not set
+ # CONFIG_TARGET_LS1021AQDS is not set
+ # CONFIG_TARGET_LS1021ATWR is not set
+ # CONFIG_TARGET_LS1043AQDS is not set
+ # CONFIG_TARGET_LS1043ARDB is not set
+ # CONFIG_TARGET_H2200 is not set
+ # CONFIG_TARGET_ZIPITZ2 is not set
+ # CONFIG_TARGET_COLIBRI_PXA270 is not set
+ # CONFIG_ARCH_UNIPHIER is not set
+ # CONFIG_STM32 is not set
+ # CONFIG_ARCH_ROCKCHIP is not set
+ # CONFIG_TARGET_THUNDERX_88XX is not set
+ # CONFIG_TARGET_SMDKV310 is not set
+ # CONFIG_TARGET_TRATS is not set
+ # CONFIG_TARGET_S5PC210_UNIVERSAL is not set
+ # CONFIG_TARGET_ORIGEN is not set
+ # CONFIG_TARGET_TRATS2 is not set
+ # CONFIG_TARGET_ODROID is not set
+ CONFIG_TARGET_ODROID_XU3=y
+ # CONFIG_TARGET_ARNDALE is not set
+ # CONFIG_TARGET_SMDK5250 is not set
+ # CONFIG_TARGET_SNOW is not set
+ # CONFIG_TARGET_SPRING is not set
+ # CONFIG_TARGET_SMDK5420 is not set
+ # CONFIG_TARGET_PEACH_PI is not set
+ # CONFIG_TARGET_PEACH_PIT is not set
+ CONFIG_SYS_MALLOC_F_LEN=0x400
+ CONFIG_SYS_MALLOC_F=y
+ CONFIG_DM_SERIAL=y
+ CONFIG_DM_SPI=y
+ CONFIG_DM_SPI_FLASH=y
+ CONFIG_DM_I2C=y
+ CONFIG_DM_GPIO=y
+ # CONFIG_ARMV7_LPAE is not set
+ 
+ #
+ # ARM debug
+ #
+ # CONFIG_DEBUG_LL is not set
+ CONFIG_DM_KEYBOARD=y
+ CONFIG_DEFAULT_DEVICE_TREE="exynos5422-odroidxu3"
+ # CONFIG_I8042_KEYB is not set
+ 
+ #
+ # General setup
+ #
+ CONFIG_LOCALVERSION=""
+ CONFIG_LOCALVERSION_AUTO=y
+ CONFIG_CC_OPTIMIZE_FOR_SIZE=y
+ CONFIG_EXPERT=y
+ # CONFIG_SYS_MALLOC_CLEAR_ON_INIT is not set
+ 
+ #
+ # Boot images
+ #
+ CONFIG_FIT=y
+ # CONFIG_FIT_VERBOSE is not set
+ # CONFIG_FIT_SIGNATURE is not set
+ CONFIG_FIT_BEST_MATCH=y
+ # CONFIG_OF_BOARD_SETUP is not set
+ # CONFIG_OF_SYSTEM_SETUP is not set
+ # CONFIG_OF_STDOUT_VIA_ALIAS is not set
+ CONFIG_SYS_EXTRA_OPTIONS=""
+ # CONFIG_SPL_LOAD_FIT is not set
+ 
+ #
+ # Boot timing
+ #
+ # CONFIG_BOOTSTAGE is not set
+ CONFIG_BOOTSTAGE_USER_COUNT=20
+ CONFIG_BOOTSTAGE_STASH_ADDR=0
+ CONFIG_BOOTSTAGE_STASH_SIZE=4096
+ # CONFIG_CONSOLE_RECORD is not set
+ 
+ #
+ # Command line interface
+ #
+ CONFIG_CMDLINE=y
+ CONFIG_HUSH_PARSER=y
+ CONFIG_SYS_HUSH_PARSER=y
+ CONFIG_SYS_PROMPT="ODROID-XU4 # "
+ 
+ #
+ # Autoboot options
+ #
+ CONFIG_AUTOBOOT_KEYED=y
+ CONFIG_AUTOBOOT_PROMPT="Autoboot in %d seconds, press <ESC> to stop\n"
+ # CONFIG_AUTOBOOT_ENCRYPTION is not set
+ CONFIG_AUTOBOOT_DELAY_STR=""
+ CONFIG_AUTOBOOT_STOP_STR="\x1b"
+ # CONFIG_AUTOBOOT_KEYED_CTRLC is not set
+ 
+ #
+ # Commands
+ #
+ 
+ #
+ # Info commands
+ #
+ CONFIG_CMD_BDI=y
+ CONFIG_CMD_CONSOLE=y
+ # CONFIG_CMD_CPU is not set
+ # CONFIG_CMD_LICENSE is not set
+ 
+ #
+ # Boot commands
+ #
+ CONFIG_CMD_BOOTD=y
+ CONFIG_CMD_BOOTM=y
+ CONFIG_CMD_BOOTZ=y
+ CONFIG_CMD_BOOTEFI=y
+ CONFIG_CMD_ELF=y
+ CONFIG_CMD_FDT=y
+ CONFIG_CMD_GO=y
+ CONFIG_CMD_RUN=y
+ CONFIG_CMD_IMI=y
+ # CONFIG_CMD_IMLS is not set
+ CONFIG_CMD_XIMG=y
+ 
+ #
+ # Environment commands
+ #
+ # CONFIG_CMD_ASKENV is not set
+ CONFIG_CMD_EXPORTENV=y
+ CONFIG_CMD_IMPORTENV=y
+ CONFIG_CMD_EDITENV=y
+ # CONFIG_CMD_GREPENV is not set
+ CONFIG_CMD_SAVEENV=y
+ CONFIG_CMD_ENV_EXISTS=y
+ 
+ #
+ # Memory commands
+ #
+ CONFIG_CMD_MEMORY=y
+ CONFIG_CMD_CRC32=y
+ # CONFIG_LOOPW is not set
+ # CONFIG_CMD_MEMTEST is not set
+ # CONFIG_CMD_MX_CYCLIC is not set
+ # CONFIG_CMD_MEMINFO is not set
+ 
+ #
+ # Device access commands
+ #
+ CONFIG_CMD_DM=y
+ # CONFIG_CMD_DEMO is not set
+ CONFIG_CMD_LOADB=y
+ CONFIG_CMD_LOADS=y
+ CONFIG_CMD_FLASH=y
+ # CONFIG_CMD_ARMFLASH is not set
+ CONFIG_CMD_MMC=y
+ # CONFIG_CMD_NAND is not set
+ # CONFIG_CMD_SF is not set
+ # CONFIG_CMD_SPI is not set
+ CONFIG_CMD_I2C=y
+ CONFIG_CMD_USB=y
+ CONFIG_CMD_DFU=y
+ CONFIG_CMD_USB_MASS_STORAGE=y
+ CONFIG_CMD_FPGA=y
+ CONFIG_CMD_GPIO=y
+ 
+ #
+ # Shell scripting commands
+ #
+ CONFIG_CMD_ECHO=y
+ CONFIG_CMD_ITEST=y
+ CONFIG_CMD_SOURCE=y
+ # CONFIG_CMD_SETEXPR is not set
+ 
+ #
+ # Network commands
+ #
+ CONFIG_CMD_NET=y
+ # CONFIG_CMD_TFTPPUT is not set
+ # CONFIG_CMD_TFTPSRV is not set
+ # CONFIG_CMD_RARP is not set
+ CONFIG_CMD_DHCP=y
+ CONFIG_CMD_NFS=y
+ CONFIG_CMD_MII=y
+ CONFIG_CMD_PING=y
+ # CONFIG_CMD_CDP is not set
+ # CONFIG_CMD_SNTP is not set
+ # CONFIG_CMD_DNS is not set
+ # CONFIG_CMD_LINK_LOCAL is not set
+ 
+ #
+ # Misc commands
+ #
+ CONFIG_CMD_CACHE=y
+ CONFIG_CMD_TIME=y
+ CONFIG_CMD_MISC=y
+ # CONFIG_CMD_TIMER is not set
+ 
+ #
+ # Power commands
+ #
+ CONFIG_CMD_PMIC=y
+ # CONFIG_CMD_REGULATOR is not set
+ 
+ #
+ # Security commands
+ #
+ 
+ #
+ # Filesystem commands
+ #
+ CONFIG_CMD_EXT2=y
+ CONFIG_CMD_EXT4=y
+ CONFIG_CMD_EXT4_WRITE=y
+ CONFIG_CMD_FAT=y
+ CONFIG_CMD_FS_GENERIC=y
+ CONFIG_SUPPORT_OF_CONTROL=y
+ 
+ #
+ # Device Tree Control
+ #
+ CONFIG_OF_CONTROL=y
+ CONFIG_OF_SEPARATE=y
+ # CONFIG_OF_EMBED is not set
+ CONFIG_NET=y
+ # CONFIG_NET_RANDOM_ETHADDR is not set
+ # CONFIG_NETCONSOLE is not set
+ CONFIG_NET_TFTP_VARS=y
+ 
+ #
+ # Device Drivers
+ #
+ 
+ #
+ # Generic Driver Options
+ #
+ CONFIG_DM=y
+ CONFIG_DM_WARN=y
+ CONFIG_DM_DEVICE_REMOVE=y
+ CONFIG_DM_STDIO=y
+ CONFIG_DM_SEQ_ALIAS=y
+ # CONFIG_SPL_DM_SEQ_ALIAS is not set
+ # CONFIG_REGMAP is not set
+ # CONFIG_SPL_REGMAP is not set
+ # CONFIG_DEVRES is not set
+ CONFIG_SIMPLE_BUS=y
+ CONFIG_OF_TRANSLATE=y
+ CONFIG_ADC=y
+ CONFIG_ADC_EXYNOS=y
+ # CONFIG_ADC_SANDBOX is not set
+ # CONFIG_BLK is not set
+ CONFIG_DISK=y
+ # CONFIG_BLOCK_CACHE is not set
+ 
+ #
+ # Clock
+ #
+ # CONFIG_CLK is not set
+ # CONFIG_CPU is not set
+ 
+ #
+ # Hardware crypto devices
+ #
+ # CONFIG_FSL_CAAM is not set
+ 
+ #
+ # Demo for driver model
+ #
+ # CONFIG_DM_DEMO is not set
+ 
+ #
+ # DFU support
+ #
+ # CONFIG_DFU_TFTP is not set
+ 
+ #
+ # DMA Support
+ #
+ # CONFIG_DMA is not set
+ # CONFIG_TI_EDMA3 is not set
+ 
+ #
+ # GPIO Support
+ #
+ # CONFIG_ALTERA_PIO is not set
+ # CONFIG_DWAPB_GPIO is not set
+ # CONFIG_ATMEL_PIO4 is not set
+ # CONFIG_INTEL_BROADWELL_GPIO is not set
+ # CONFIG_LPC32XX_GPIO is not set
+ # CONFIG_MSM_GPIO is not set
+ # CONFIG_ROCKCHIP_GPIO is not set
+ # CONFIG_VYBRID_GPIO is not set
+ 
+ #
+ # I2C support
+ #
+ CONFIG_DM_I2C_COMPAT=y
+ # CONFIG_DM_I2C_GPIO is not set
+ # CONFIG_SYS_I2C_INTEL is not set
+ # CONFIG_SYS_I2C_ROCKCHIP is not set
+ # CONFIG_I2C_MUX is not set
+ # CONFIG_CROS_EC_KEYB is not set
+ 
+ #
+ # LED Support
+ #
+ # CONFIG_LED is not set
+ 
+ #
+ # Memory Controller drivers
+ #
+ 
+ #
+ # Multifunction device drivers
+ #
+ # CONFIG_MISC is not set
+ # CONFIG_CROS_EC is not set
+ # CONFIG_FSL_SEC_MON is not set
+ # CONFIG_MXC_OCOTP is not set
+ # CONFIG_PWRSEQ is not set
+ # CONFIG_PCA9551_LED is not set
+ # CONFIG_RESET is not set
+ # CONFIG_WINBOND_W83627 is not set
+ 
+ #
+ # MMC Host controller Support
+ #
+ # CONFIG_DM_MMC is not set
+ 
+ #
+ # MTD Support
+ #
+ # CONFIG_MTD is not set
+ 
+ #
+ # NAND Device Support
+ #
+ # CONFIG_NAND_DENALI is not set
+ # CONFIG_NAND_VF610_NFC is not set
+ # CONFIG_NAND_PXA3XX is not set
+ # CONFIG_NAND_ARASAN is not set
+ 
+ #
+ # Generic NAND options
+ #
+ 
+ #
+ # SPI Flash Support
+ #
+ # CONFIG_SPI_FLASH is not set
+ # CONFIG_DM_ETH is not set
+ # CONFIG_PHYLIB is not set
+ # CONFIG_NETDEVICES is not set
+ 
+ #
+ # PCI
+ #
+ # CONFIG_DM_PCI is not set
+ 
+ #
+ # Pin controllers
+ #
+ # CONFIG_PINCTRL is not set
+ 
+ #
+ # Power
+ #
+ CONFIG_DM_PMIC=y
+ CONFIG_PMIC_CHILDREN=y
+ CONFIG_SPL_PMIC_CHILDREN=y
+ # CONFIG_PMIC_ACT8846 is not set
+ # CONFIG_DM_PMIC_PFUZE100 is not set
+ # CONFIG_DM_PMIC_MAX77686 is not set
+ # CONFIG_PMIC_PM8916 is not set
+ # CONFIG_PMIC_RK808 is not set
+ CONFIG_PMIC_S2MPS11=y
+ # CONFIG_DM_PMIC_SANDBOX is not set
+ # CONFIG_PMIC_S5M8767 is not set
+ # CONFIG_PMIC_TPS65090 is not set
+ CONFIG_DM_REGULATOR=y
+ # CONFIG_SPL_DM_REGULATOR is not set
+ # CONFIG_DM_REGULATOR_FIXED is not set
+ # CONFIG_DM_PWM is not set
+ # CONFIG_RAM is not set
+ 
+ #
+ # Remote Processor drivers
+ #
+ 
+ #
+ # Real Time Clock
+ #
+ # CONFIG_DM_RTC is not set
+ 
+ #
+ # Serial drivers
+ #
+ CONFIG_REQUIRE_SERIAL_CONSOLE=y
+ CONFIG_SERIAL_PRESENT=y
+ CONFIG_SPL_SERIAL_PRESENT=y
+ # CONFIG_DEBUG_UART is not set
+ # CONFIG_DEBUG_UART_SKIP_INIT is not set
+ # CONFIG_ALTERA_JTAG_UART is not set
+ # CONFIG_ALTERA_UART is not set
+ # CONFIG_FSL_LPUART is not set
+ # CONFIG_SYS_NS16550 is not set
+ # CONFIG_MSM_SERIAL is not set
+ 
+ #
+ # Sound support
+ #
+ # CONFIG_SOUND is not set
+ 
+ #
+ # SPI Support
+ #
+ # CONFIG_ALTERA_SPI is not set
+ # CONFIG_CADENCE_QSPI is not set
+ # CONFIG_DESIGNWARE_SPI is not set
+ # CONFIG_EXYNOS_SPI is not set
+ # CONFIG_FSL_DSPI is not set
+ # CONFIG_FSL_QSPI is not set
+ # CONFIG_ICH_SPI is not set
+ # CONFIG_ROCKCHIP_SPI is not set
+ # CONFIG_TEGRA114_SPI is not set
+ # CONFIG_TEGRA20_SFLASH is not set
+ # CONFIG_TEGRA20_SLINK is not set
+ # CONFIG_TEGRA210_QSPI is not set
+ # CONFIG_XILINX_SPI is not set
+ # CONFIG_OMAP3_SPI is not set
+ # CONFIG_FSL_ESPI is not set
+ # CONFIG_TI_QSPI is not set
+ 
+ #
+ # SPMI support
+ #
+ # CONFIG_SPMI is not set
+ # CONFIG_DM_THERMAL is not set
+ 
+ #
+ # Timer Support
+ #
+ # CONFIG_TIMER is not set
+ 
+ #
+ # TPM support
+ #
+ CONFIG_USB=y
+ CONFIG_DM_USB=y
+ 
+ #
+ # USB Host Controller Drivers
+ #
+ # CONFIG_USB_XHCI_HCD is not set
+ # CONFIG_USB_XHCI is not set
+ # CONFIG_USB_OHCI_GENERIC is not set
+ # CONFIG_USB_EHCI_HCD is not set
+ # CONFIG_USB_EHCI is not set
+ CONFIG_USB_DWC3=y
+ # CONFIG_USB_DWC3_HOST is not set
+ CONFIG_USB_DWC3_GADGET=y
+ 
+ #
+ # Platform Glue Driver Support
+ #
+ # CONFIG_USB_DWC3_OMAP is not set
+ 
+ #
+ # PHY Subsystem
+ #
+ # CONFIG_USB_DWC3_PHY_OMAP is not set
+ CONFIG_USB_DWC3_PHY_SAMSUNG=y
+ 
+ #
+ # MUSB Controller Driver
+ #
+ # CONFIG_USB_MUSB_HOST is not set
+ # CONFIG_USB_MUSB_GADGET is not set
+ 
+ #
+ # ULPI drivers
+ #
+ 
+ #
+ # USB peripherals
+ #
+ # CONFIG_USB_STORAGE is not set
+ # CONFIG_USB_KEYBOARD is not set
+ CONFIG_USB_GADGET=y
+ # CONFIG_USB_GADGET_ATMEL_USBA is not set
+ # CONFIG_USB_GADGET_DWC2_OTG is not set
+ # CONFIG_CI_UDC is not set
+ CONFIG_USB_GADGET_VBUS_DRAW=2
+ CONFIG_USB_GADGET_DUALSPEED=y
+ CONFIG_USB_GADGET_DOWNLOAD=y
+ CONFIG_G_DNL_MANUFACTURER="Samsung"
+ CONFIG_G_DNL_VENDOR_NUM=0x04e8
+ CONFIG_G_DNL_PRODUCT_NUM=0x6601
+ 
+ #
+ # Graphics support
+ #
+ # CONFIG_DM_VIDEO is not set
+ 
+ #
+ # TrueType Fonts
+ #
+ # CONFIG_VIDEO_VESA is not set
+ # CONFIG_VIDEO_LCD_ANX9804 is not set
+ # CONFIG_VIDEO_LCD_SSD2828 is not set
+ # CONFIG_VIDEO_MVEBU is not set
+ # CONFIG_I2C_EDID is not set
+ # CONFIG_DISPLAY is not set
+ # CONFIG_VIDEO_TEGRA20 is not set
+ # CONFIG_VIDEO_BRIDGE is not set
+ # CONFIG_PHYS_TO_BUS is not set
+ 
+ #
+ # File systems
+ #
+ 
+ #
+ # Library routines
+ #
+ # CONFIG_CC_OPTIMIZE_LIBS_FOR_SPEED is not set
+ CONFIG_HAVE_PRIVATE_LIBGCC=y
+ # CONFIG_USE_PRIVATE_LIBGCC is not set
+ CONFIG_SYS_HZ=1000
+ # CONFIG_USE_TINY_PRINTF is not set
+ CONFIG_REGEX=y
+ # CONFIG_LIB_RAND is not set
+ # CONFIG_CMD_DHRYSTONE is not set
+ # CONFIG_RSA is not set
+ # CONFIG_TPM is not set
+ 
+ #
+ # Hashing Support
+ #
+ # CONFIG_SHA1 is not set
+ # CONFIG_SHA256 is not set
+ # CONFIG_SHA_HW_ACCEL is not set
+ 
+ #
+ # Compression Support
+ #
+ # CONFIG_LZ4 is not set
+ CONFIG_ERRNO_STR=y
+ CONFIG_OF_LIBFDT=y
+ # CONFIG_SPL_OF_LIBFDT is not set
+ CONFIG_EFI_LOADER=y
+ # CONFIG_UNIT_TEST is not set
