--- a/arch/arm/boot/dts/Makefile        2018-04-30 14:37:29.138206953 +0200
+++ b/arch/arm/boot/dts/Makefile        2018-04-30 14:37:46.264298398 +0200
@@ -769,7 +769,8 @@
        rk3288-veyron-mickey.dtb \
        rk3288-veyron-minnie.dtb \
        rk3288-veyron-pinky.dtb \
-       rk3288-veyron-speedy.dtb
+       rk3288-veyron-speedy.dtb \
+       rk3288-xt-q8-v10.dtb
 dtb-$(CONFIG_ARCH_S3C24XX) += \
        s3c2416-smdk2416.dtb
 dtb-$(CONFIG_ARCH_S3C64XX) += \
                                                                                                                                                                                                                            1012,2        Fon

