diff --git a/arch/arm/boot/dts/zynq-microzed.dts b/arch/arm/boot/dts/zynq-microzed.dts
index 6ed84fb15..4d7ba65c2 100644
--- a/arch/arm/boot/dts/zynq-microzed.dts
+++ b/arch/arm/boot/dts/zynq-microzed.dts
@@ -29,6 +29,15 @@
 		compatible = "usb-nop-xceiv";
 		#phy-cells = <0>;
 	};
+
+	leds {
+		compatible = "gpio-leds";
+		status {
+			label = "status";
+			gpios = <&gpio0 0x2f 0x0>;
+			linux,default-trigger = "heartbeat";
+		};
+	};
 };
 
 &clkc {
