diff -ruN a/arch/arm64/boot/dts/rockchip/overlay/rk3566-i2c4.dts b/arch/arm64/boot/dts/rockchip/overlay/rk3566-i2c4.dts
--- a/arch/arm64/boot/dts/rockchip/overlay/rk3566-i2c4.dts	1970-01-01 08:00:00.000000000 +0800
+++ b/arch/arm64/boot/dts/rockchip/overlay/rk3566-i2c4.dts	2024-08-01 10:43:17.000000000 +0800
@@ -0,0 +1,18 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/rockchip.h>
+#include <dt-bindings/power/rk3568-power.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {
+		
+	fragment@0 {
+		target = <&i2c4>;
+
+		__overlay__ {
+			status = "okay";			
+		};
+	};	
+};
\ No newline at end of file
diff -ruN a/arch/arm64/boot/dts/rockchip/overlay/rk3566-mcp2515.dts b/arch/arm64/boot/dts/rockchip/overlay/rk3566-mcp2515.dts
--- a/arch/arm64/boot/dts/rockchip/overlay/rk3566-mcp2515.dts	1970-01-01 08:00:00.000000000 +0800
+++ b/arch/arm64/boot/dts/rockchip/overlay/rk3566-mcp2515.dts	2024-08-01 10:43:17.000000000 +0800
@@ -0,0 +1,28 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/rockchip.h>
+#include <dt-bindings/power/rk3568-power.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {
+		
+	fragment@0 {
+		target = <&spi1>;
+
+		__overlay__ {
+			status = "okay";
+			//pinctrl-0=<&spi1m1_pins>;
+		};
+	};
+
+	fragment@1 {
+		target = <&can_mcp2515>;
+
+		__overlay__ {
+			 status = "okay";
+		};
+	};
+		
+};
\ No newline at end of file
diff -ruN a/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm0.dts b/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm0.dts
--- a/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm0.dts	1970-01-01 08:00:00.000000000 +0800
+++ b/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm0.dts	2024-08-01 10:43:17.000000000 +0800
@@ -0,0 +1,19 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/rockchip.h>
+#include <dt-bindings/power/rk3568-power.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {		
+	fragment@0 {
+		target = <&pwm0>;
+
+		__overlay__ {			
+			status = "okay";
+		
+		};
+	};
+	
+};	
diff -ruN a/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm1.dts b/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm1.dts
--- a/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm1.dts	1970-01-01 08:00:00.000000000 +0800
+++ b/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm1.dts	2024-08-01 10:43:17.000000000 +0800
@@ -0,0 +1,20 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/rockchip.h>
+#include <dt-bindings/power/rk3568-power.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {		
+	fragment@0 {
+		target = <&pwm1>;
+
+		__overlay__ {
+			pinctrl-names = "default";			
+			status = "okay";
+		
+		};
+	};
+	
+};	
