diff -ruN a/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm13.dts b/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm13.dts
--- a/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm13.dts	1970-01-01 08:00:00.000000000 +0800
+++ b/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm13.dts	2024-08-01 10:43:17.000000000 +0800
@@ -0,0 +1,20 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/rockchip.h>
+#include <dt-bindings/power/rk3568-power.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {		
+	fragment@0 {
+		target = <&pwm13>;
+
+		__overlay__ {
+			pinctrl-names = "default";
+			status = "okay";
+		
+		};
+	};
+	
+};	
diff -ruN a/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm14.dts b/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm14.dts
--- a/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm14.dts	1970-01-01 08:00:00.000000000 +0800
+++ b/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm14.dts	2024-08-01 10:43:17.000000000 +0800
@@ -0,0 +1,19 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/rockchip.h>
+#include <dt-bindings/power/rk3568-power.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {		
+	fragment@0 {
+		target = <&pwm14>;
+
+		__overlay__ {
+			status = "okay";
+		
+		};
+	};
+	
+};	
diff -ruN a/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm15.dts b/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm15.dts
--- a/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm15.dts	1970-01-01 08:00:00.000000000 +0800
+++ b/arch/arm64/boot/dts/rockchip/overlay/rk3566-pwm15.dts	2024-08-01 10:43:17.000000000 +0800
@@ -0,0 +1,19 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/rockchip.h>
+#include <dt-bindings/power/rk3568-power.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {		
+	fragment@0 {
+		target = <&pwm15>;
+
+		__overlay__ {
+			status = "okay";
+		
+		};
+	};
+	
+};	
diff -ruN a/arch/arm64/boot/dts/rockchip/overlay/rk3566-rtc-pcf8563.dts b/arch/arm64/boot/dts/rockchip/overlay/rk3566-rtc-pcf8563.dts
--- a/arch/arm64/boot/dts/rockchip/overlay/rk3566-rtc-pcf8563.dts	1970-01-01 08:00:00.000000000 +0800
+++ b/arch/arm64/boot/dts/rockchip/overlay/rk3566-rtc-pcf8563.dts	2024-08-01 10:43:17.000000000 +0800
@@ -0,0 +1,19 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/rockchip.h>
+#include <dt-bindings/power/rk3568-power.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {		
+	fragment@0 {
+		target = <&rtc_pcf8563>;
+
+		__overlay__ {
+			
+			status = "okay";		
+		};
+	};
+	
+};	
\ No newline at end of file
