diff --git a/include/configs/odroid_xu4.h b/include/configs/odroid_xu4.h
index 29ddaea..a1fa0df 100755
--- a/include/configs/odroid_xu4.h
+++ b/include/configs/odroid_xu4.h
@@ -243,7 +243,7 @@
 	"copy_uboot_emmc2sd="UBOOT_COPY_EMMC2SD
 
 #undef CONFIG_BOOTCOMMAND
-#define CONFIG_BOOTCOMMAND "cfgload;movi r k 0 40008000;bootz 40008000"
+#define CONFIG_BOOTCOMMAND "cfgload;run distro_bootcmd;movi r k 0 40008000;bootz 40008000"
 
 #undef CONFIG_BOOTARGS
 /* Android Default bootargs */
