diff --git a/arch/arm/boot/dts/sun8i-h3.dtsi b/arch/arm/boot/dts/sun8i-h3.dtsi
index 12dc849a..d1b4cd79 100644
--- a/arch/arm/boot/dts/sun8i-h3.dtsi
+++ b/arch/arm/boot/dts/sun8i-h3.dtsi
@@ -509,7 +523,7 @@
 			};

 			uart3_pins: uart3 {
-				allwinner,pins = "PA13", "PA14";
+				allwinner,pins = "PA13", "PA14", "PA15", "PA16";
 				allwinner,function = "uart3";
 				allwinner,drive = <SUN4I_PINCTRL_10_MA>;
 				allwinner,pull = <SUN4I_PINCTRL_NO_PULL>;
