diff --git a/arch/arm/boot/dts/sunxi-h3-h5.dtsi b/arch/arm/boot/dts/sunxi-h3-h5.dtsi
index c3e2226..37eab9b 100644
--- a/arch/arm/boot/dts/sunxi-h3-h5.dtsi
+++ b/arch/arm/boot/dts/sunxi-h3-h5.dtsi
@@ -43,9 +43,11 @@
 #include <dt-bindings/clock/sun8i-h3-ccu.h>
 #include <dt-bindings/interrupt-controller/arm-gic.h>
 #include <dt-bindings/reset/sun8i-h3-ccu.h>
+#include <dt-bindings/pinctrl/sun4i-a10.h>
+#include <dt-bindings/gpio/gpio.h>
 
 / {
 	interrupt-parent = <&gic>;
 	#address-cells = <1>;
 	#size-cells = <1>;
 
@@ -735,6 +773,14 @@
 				pins = "PL0", "PL1";
 				function = "s_twi";
 			};
+
+			usb0_vbus_pin_opipc: usb0_vbus_pin@0 {
+				allwinner,pins = "PL2";
+				allwinner,function = "gpio_out";
+				allwinner,drive = <SUN4I_PINCTRL_10_MA>;
+				allwinner,pull = <SUN4I_PINCTRL_NO_PULL>;
+			};
+
 		};
 	};
 
@@ -745,4 +791,13 @@
 			thermal-sensors = <&ths 0>;
 		};
 	};
+
+	reg_usb0_vbus {
+		/delete-property/ pinctrl-0;
+		pinctrl-0 = <&usb0_vbus_pin_opipc>;
+		/delete-property/ gpio;
+		gpio = <&r_pio 0 2 GPIO_ACTIVE_HIGH>; /* PL2 */
+		status = "disabled";
+	};
+
 };
