From 497b02d0f2a79f5af4648adb232e011c6a3032bf Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megous@megous.com>
Date: Thu, 17 Sep 2020 00:11:15 +0200
Subject: [PATCH 187/351] misc: modem-power: Whitespace

Signed-off-by: Ondrej Jirman <megous@megous.com>
---
 drivers/misc/modem-power.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/misc/modem-power.c b/drivers/misc/modem-power.c
index aee4479b0d6a..d80eb9dacf7b 100644
--- a/drivers/misc/modem-power.c
+++ b/drivers/misc/modem-power.c
@@ -317,7 +317,7 @@ static const struct mpwr_eg25_qcfg mpwr_eg25_qcfgs[] = {
 	{ "airplanecontrol",    "1",   mpwr_eg25_qcfg_airplanecontrol_is_ok },
 
 	// available since firmware R07A08_01.002.01.002
-	{ "fast/poweroff", "1" },
+	{ "fast/poweroff", 	"1" },
 };
 
 static char* mpwr_serdev_get_response_value(struct mpwr_dev *mpwr,
-- 
2.34.0

