diff --git a/include/configs/mvebu_armada-common.h b/include/configs/mvebu_armada-common.h
old mode 100644
new mode 100755
index c406a7a..ddd7f5e
--- a/include/configs/mvebu_armada-common.h
+++ b/include/configs/mvebu_armada-common.h
@@ -36,7 +36,7 @@
 #define CONFIG_ETHPRIME			"eth0"
 #define CONFIG_ROOTPATH                 "/srv/nfs/" /* Default Dir for NFS */
 #define CONFIG_EXTRA_ENV_SETTINGS	"kernel_addr=0x5000000\0"	\
-					"initrd_addr=0xa00000\0"	\
+					"initrd_addr=0x1100000\0"	\
 					"initrd_size=0x2000000\0"	\
 					"fdt_addr=0x4f00000\0"		\
 					"loadaddr=0x5000000\0"		\
@@ -50,7 +50,8 @@
 					"eth1addr=00:51:82:11:22:01\0"	\
 					"eth2addr=00:51:82:11:22:02\0"	\
 					"eth3addr=00:51:82:11:22:03\0"	\
-					"image_name=Image\0"		\
+					"image_name=boot/Image\0"		\
+					"load_script=if test -e mmc 0:1 boot/boot.scr; then echo \"... booting from SD\";setenv boot_interface mmc;else echo \"... booting from USB/SATA\";usb start;setenv boot_interface usb;fi;if test -e $boot_interface 0:1 boot/boot.scr;then ext4load $boot_interface 0:1 0x00800000 boot/boot.scr; source; fi\0" \
 					"get_ramfs=if test \"${ramfs_name}\"" \
 						" != \"-\"; then setenv " \
 						"ramfs_addr 0x8000000; " \
@@ -69,7 +70,7 @@
 						":$netdev:none nfsroot="\
 						"$serverip:$rootpath " \
 						"$extra_params"
-#define CONFIG_BOOTCOMMAND	"run get_images; run set_bootargs; " \
+#define CONFIG_BOOTCOMMAND	"run get_images; run set_bootargs; run load_script;" \
 				"booti $kernel_addr $ramfs_addr $fdt_addr"
 #define CONFIG_ENV_OVERWRITE	/* ethaddr can be reprogrammed */
 /*
