From 2bd714126a251fe47148b610b2b4effeca08c894 Mon Sep 17 00:00:00 2001
From: Patrick Yavitz <pyavitz@armbian.com>
Date: Wed, 30 Oct 2024 05:44:36 -0400
Subject: [PATCH] mmc: sunxi_mmc: bpi-m4-zero emmc boot support

Signed-off-by: Patrick Yavitz <pyavitz@armbian.com>
---
 drivers/mmc/sunxi_mmc.c | 18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

diff --git a/drivers/mmc/sunxi_mmc.c b/drivers/mmc/sunxi_mmc.c
index 0b56d1405b..d4a03f747c 100644
--- a/drivers/mmc/sunxi_mmc.c
+++ b/drivers/mmc/sunxi_mmc.c
@@ -542,6 +542,10 @@ struct mmc *sunxi_mmc_init(int sdc_no)
 
 	cfg->f_min = 400000;
 	cfg->f_max = 52000000;
+	
+	if (sdc_no == 2) {
+		cfg->f_max = 8000000;
+	}
 
 	if (mmc_resource_init(sdc_no) != 0)
 		return NULL;
@@ -639,20 +643,26 @@ static int sunxi_mmc_probe(struct udevice *dev)
 	struct mmc_config *cfg = &plat->cfg;
 	struct ofnode_phandle_args args;
 	u32 *ccu_reg;
-	int ret;
+	int bus_width, ret;
 
 	cfg->name = dev->name;
+	bus_width = dev_read_u32_default(dev, "bus-width", 1);
 
 	cfg->voltages = MMC_VDD_32_33 | MMC_VDD_33_34;
+	cfg->host_caps = 0;
+	if (bus_width == 8)
+		cfg->host_caps |= MMC_MODE_8BIT;
+	if (bus_width >= 4)
+		cfg->host_caps |= MMC_MODE_4BIT;
 	cfg->host_caps = MMC_MODE_HS_52MHz | MMC_MODE_HS;
 	cfg->b_max = CONFIG_SYS_MMC_MAX_BLK_COUNT;
 
 	cfg->f_min = 400000;
 	cfg->f_max = 52000000;
 
-	ret = mmc_of_parse(dev, cfg);
-	if (ret)
-		return ret;
+	if (bus_width == 8) {
+		cfg->f_max = 52000000;
+	}
 
 	priv->reg = dev_read_addr_ptr(dev);
 
-- 
2.39.5

