From ad10cf66b98acbdc21f431e262633e18b2e93c70 Mon Sep 17 00:00:00 2001
From: Kali Prasad <kprasadvnsi@protonmail.com>
Date: Sun, 24 Nov 2024 07:51:12 -0500
Subject: [PATCH] sunxi: h616 ths workaround

Signed-off-by: Kali Prasad <kprasadvnsi@protonmail.com>
---
 board/sunxi/board.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/board/sunxi/board.c b/board/sunxi/board.c
index 961cdcde74..af2b372a56 100644
--- a/board/sunxi/board.c
+++ b/board/sunxi/board.c
@@ -225,6 +225,15 @@ int board_init(void)
 	if (ret)
 		return ret;
 
+#if CONFIG_MACH_SUN50I_H616
+	/*
+	 * The bit[16] of register reg[0x03000000] must be zero for the THS
+	 * driver to work properly in the kernel. The BSP u-boot is putting
+	 * the whole register to zero so we are doing the same.
+	 */
+	writel(0x0, SUNXI_SRAMC_BASE);
+#endif
+
 	eth_init_board();
 
 	return 0;
-- 
2.39.5

