diff --git a/arch/arm/dts/meson-gx.dtsi b/arch/arm/dts/meson-gx.dtsi
index f1e5cdba..221923c1 100644
--- a/arch/arm/dts/meson-gx.dtsi
+++ b/arch/arm/dts/meson-gx.dtsi
@@ -181,6 +181,8 @@
 		#size-cells = <2>;
 		ranges;
 
+		u-boot,dm-spl;
+
 		cbus: bus@c1100000 {
 			compatible = "simple-bus";
 			reg = <0x0 0xc1100000 0x0 0x100000>;
@@ -505,6 +507,7 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
 			amlogic,canvas = <&canvas>;
+			u-boot,dm-spl;
 
 			/* CVBS VDAC output port */
 			cvbs_vdac_port: port@0 {
@@ -523,7 +526,9 @@
 
 		hdmi_tx: hdmi-tx@c883a000 {
 			compatible = "amlogic,meson-gx-dw-hdmi";
-			reg = <0x0 0xc883a000 0x0 0x1c>;
+			reg = <0x0 0xc883a000 0x0 0x1c>,
+			      <0x0 0xc883c000 0x0 0x1000>;
+			reg-names = "hdmitx", "hhi";
 			interrupts = <GIC_SPI 57 IRQ_TYPE_EDGE_RISING>;
 			#address-cells = <1>;
 			#size-cells = <0>;
diff --git a/arch/arm/dts/meson-gxl-s905x-libretech-cc.dts b/arch/arm/dts/meson-gxl-s905x-libretech-cc.dts
index a23252ef..c92b67a6 100644
--- a/arch/arm/dts/meson-gxl-s905x-libretech-cc.dts
+++ b/arch/arm/dts/meson-gxl-s905x-libretech-cc.dts
@@ -54,16 +54,18 @@
 		compatible = "gpio-leds";
 
 		led-system {
-			label = "librecomputer:system-status";
+			label = "librecomputer:green:disk";
 			gpios = <&gpio GPIODV_24 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "heartbeat";
 			default-state = "on";
 			panic-indicator;
 		};
 
 		led-blue {
-			label = "librecomputer:blue";
+			label = "librecomputer:blue:cpu";
 			gpios = <&gpio_ao GPIOAO_2 GPIO_ACTIVE_HIGH>;
-			linux,default-trigger = "heartbeat";
+			linux,default-trigger = "activity";
+			default-state = "off";
 		};
 	};
 
