diff --git a/u-boot-pine64/include/configs/sun50iw1p1.h b/u-boot-pine64/include/configs/sun50iw1p1.h
index 563f4df..1e62081 100644
--- a/u-boot-pine64/include/configs/sun50iw1p1.h
+++ b/u-boot-pine64/include/configs/sun50iw1p1.h
@@ -350,7 +350,7 @@
 	"initrd_filename=initrd.img\0" \
 	"bootenv_filename=uEnv.txt\0" \
 	"load_bootenv=" \
-		"fatload mmc 0:1 ${load_addr} ${bootenv_filename}\0" \
+		"ext4load mmc 0:1 ${load_addr} /boot/${bootenv_filename} || fatload mmc 0:1 ${load_addr} ${bootenv_filename} || ext4load mmc 0:1 ${load_addr} /boot/${bootenv_filename}\0" \
 	"import_bootenv=" \
 		"env import -t ${load_addr} ${filesize}\0" \
 	"load_dtb=" \
@@ -366,7 +366,7 @@
 		"fatload mmc 0:1 ${initrd_addr} ${initrd_filename}; "\
 		"setenv initrd_size ${filesize}\0" \
 	"load_bootscript=" \
-		"fatload mmc 0:1 ${load_addr} ${script}\0" \
+		"ext4load mmc 0:1 ${load_addr} /boot/${script} || fatload mmc 0:1 ${load_addr} ${script} || ext4load mmc 0:1 ${load_addr} ${script}\0" \
 	"scriptboot=source ${load_addr}\0" \
 	"set_cmdline=" \
 		"setenv bootargs console=${console} ${optargs} " \
