--- a/arch/arm64/boot/dts/rockchip/rk3399-firefly.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3399-firefly.dts
@@ -32,6 +32,8 @@
 		pinctrl-names = "default";
 	};
 
+	/delete-node/ leds;
+
 	leds {
 		compatible = "gpio-leds";
 		pinctrl-names = "default";
@@ -72,6 +74,17 @@
 			sound-dai = <&rt5640>;
 		};
 	};
+	
+	vcc3v3_3g: vcc3v3-3g-regulator {
+		compatible = "regulator-fixed";
+		enable-active-high;
+		gpio = <&gpio0 2 GPIO_ACTIVE_HIGH>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&pcie_3g_drv>;
+		regulator-name = "vcc3v3_3g";
+		regulator-always-on;
+		regulator-boot-on;
+	};
 
 };
 
@@ -209,6 +222,8 @@
 		};
 	};
 
+	/delete-node/ leds;
+
 	leds {
 		work_led_pin: work-led-pin {
 			rockchip,pins = <2 RK_PD3 RK_FUNC_GPIO &pcfg_pull_none>;
@@ -220,7 +235,6 @@
 	};
 };
 
-
 &sdio0 {
 	/* WiFi & BT combo module Ampak AP6356S */
 	bus-width = <4>;
