new file mode 100644
index 000000000000..b85508c12742
--- /dev/null
+++ b/arch/arm64/boot/dts/rockchip/rk3399-roc-pc-plus.dts
@@ -0,0 +1,33 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2017 T-Chip Intelligent Technology Co., Ltd
+ */
+
+/dts-v1/;
+#include "rk3399-roc-pc-mezzanine.dts"
+
+/ {
+	model = "Firefly roc-rk3399-pc PLUS";
+	compatible = "firefly,roc-rk3399-pc", "rockchip,rk3399";
+};
+
+&rk808{
+	rtc {
+		compatible = "rk808-rtc";	
+		status = "disabled";
+	};
+};
+
+&i2c0 {
+	status = "okay";
+	hym8563: hym8563@51 {
+		compatible = "haoyu,hym8563";
+		reg = <0x51>;
+		interrupt-parent = <&gpio0>;
+		interrupts = <5 IRQ_TYPE_EDGE_FALLING>;
+		pinctrl-names = "default";
+		#clock-cells = <0>;
+		clock-frequency = <32768>;
+		clock-output-names = "xin32k";
+	};
+};
