diff --git a/include/configs/rv1103b_common.h b/include/configs/rv1103b_common.h
index f6ee6f88dc..32854c2af4 100644
--- a/include/configs/rv1103b_common.h
+++ b/include/configs/rv1103b_common.h
@@ -10,7 +10,8 @@
 #include "rockchip-common.h"
 
 #define COUNTER_FREQUENCY		24000000
-#define CONFIG_SYS_MALLOC_LEN		(16 << 20)
+#define CONFIG_SYS_MALLOC_LEN		(8 << 20)
+#define CONFIG_SYS_MEM_TOP_HIDE		0x00300000
 #define CONFIG_SYS_CBSIZE		1024
 #define CONFIG_SYS_NS16550_MEM32
 
diff --git a/include/configs/rv1106_common.h b/include/configs/rv1106_common.h
index dc927a8f36..0f062426b1 100644
--- a/include/configs/rv1106_common.h
+++ b/include/configs/rv1106_common.h
@@ -12,7 +12,8 @@
 #include "rockchip-common.h"
 
 #define COUNTER_FREQUENCY		24000000
-#define CONFIG_SYS_MALLOC_LEN		(16 << 20)
+#define CONFIG_SYS_MALLOC_LEN		(8 << 20)
+#define CONFIG_SYS_MEM_TOP_HIDE		0x00300000
 #define CONFIG_SYS_CBSIZE		1024
 #define CONFIG_SYS_NS16550_MEM32
 
