diff --git a/include/configs/rv1103b_common.h b/include/configs/rv1103b_common.h
index f6ee6f88dc..f30d0197e0 100644
--- a/include/configs/rv1103b_common.h
+++ b/include/configs/rv1103b_common.h
@@ -57,19 +57,14 @@
 #ifdef CONFIG_ENV_MEM_LAYOUT
 #define ENV_MEM_LAYOUT_SETTINGS		CONFIG_ENV_MEM_LAYOUT_SETTINGS
 #else
-/*
- *   Image:  0 - 8M
- *  zImage:  8 - 12M
- *     fdt: 12 - 13M
- * ramdisk: 14 ...
- */
 #define ENV_MEM_LAYOUT_SETTINGS		\
-	"scriptaddr=0x00b00000\0"	\
-	"pxefile_addr_r=0x00c00000\0"	\
-	"fdt_addr_r=0x00c00000\0"	\
-	"kernel_addr_c=0x00808000\0"	\
-	"kernel_addr_r=0x00008000\0"	\
-	"ramdisk_addr_r=0x000e00000\0"
+	"scriptaddr=0x02000000\0"	\
+	"fdt_addr_r=0x03800000\0"	\
+	"kernel_addr_r=0x00800000\0"	\
+	"ramdisk_addr_r=0x02400000\0"	\
+	"fdt_high=0x03900000\0"	\
+	"initrd_high=0x03900000\0"	\
+	"load_addr=0x02000000\0"
 #endif
 
 #define CONFIG_EXTRA_ENV_SETTINGS	\
diff --git a/include/configs/rv1106_common.h b/include/configs/rv1106_common.h
index dc927a8f36..d6912a88ba 100644
--- a/include/configs/rv1106_common.h
+++ b/include/configs/rv1106_common.h
@@ -67,19 +67,14 @@
 #ifdef CONFIG_ENV_MEM_LAYOUT
 #define ENV_MEM_LAYOUT_SETTINGS		CONFIG_ENV_MEM_LAYOUT_SETTINGS
 #else
-/*
- *   Image:  0 - 8M
- *  zImage:  8 - 12M
- *     fdt: 12 - 13M
- * ramdisk: 14 ...
- */
 #define ENV_MEM_LAYOUT_SETTINGS		\
-	"scriptaddr=0x00b00000\0"	\
-	"pxefile_addr_r=0x00c00000\0"	\
-	"fdt_addr_r=0x00c00000\0"	\
-	"kernel_addr_c=0x00808000\0"	\
-	"kernel_addr_r=0x00008000\0"	\
-	"ramdisk_addr_r=0x000e00000\0"
+	"scriptaddr=0x02000000\0"	\
+	"fdt_addr_r=0x03800000\0"	\
+	"kernel_addr_r=0x00800000\0"	\
+	"ramdisk_addr_r=0x02400000\0"	\
+	"fdt_high=0x03900000\0"	\
+	"initrd_high=0x03900000\0"	\
+	"load_addr=0x02000000\0"
 #endif
 
 #define CONFIG_EXTRA_ENV_SETTINGS	\
