diff --git linux-4.18.7/arch/arm64/boot/dts/marvell/armada-3720-espressobin.dts b/arch/arm64/boot/dts/marvell/armada-3720-espressobin.dts
index 3ab25ad40..0f99e751c 100644
--- linux-4.18.7/arch/arm64/boot/dts/marvell/armada-3720-espressobin.dts
+++ b/arch/arm64/boot/dts/marvell/armada-3720-espressobin.dts
@@ -41,6 +41,15 @@
 			  3300000 0x0>;
 		enable-active-high;
 	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		red {
+			label = "espressobin:red:usr";
+			gpios = <&gpionb 2 GPIO_ACTIVE_LOW>;
+		};
+	};
 };
 
 /* J9 */
@@ -68,9 +77,17 @@
 
 	flash@0 {
 		reg = <0>;
-		compatible = "winbond,w25q32dw", "jedec,spi-flash";
+		/*
+		 * Originally "winbond,w25q32dw", but since the manufacturer is known
+		 * to have replaced the part with "macronix,mx25u3235f" in some board
+		 * batches, just use the generic "jedec,spi-nor" and let the actual
+		 * chip type be probed. The partition table still depends on the chip
+		 * being 4 MiB in size.
+		 */
+		compatible = "jedec,spi-nor";
 		spi-max-frequency = <104000000>;
 		m25p,fast-read;
+		status = "okay";
 
 		partitions {
 			compatible = "fixed-partitions";
@@ -79,12 +96,12 @@
 
 			partition@0 {
 				label = "uboot";
-				reg = <0 0x180000>;
+				reg = <0 0x3f0000>;
 			};
 
-			partition@180000 {
+			partition@3f0000 {
 				label = "ubootenv";
-				reg = <0x180000 0x10000>;
+				reg = <0x3f0000 0x10000>;
 			};
 		};
 	};
