From 59cc95913fc1917736eeb52b823eb20d86d37911 Mon Sep 17 00:00:00 2001
From: Alex Ling <ling_kasim@hotmail.com>
Date: Fri, 23 Jan 2026 09:31:36 +0800
Subject: [PATCH 10/15] drivers: power: Fix name for sc8280xp battery

Signed-off-by: Alex Ling <ling_kasim@hotmail.com>
---
 drivers/power/supply/qcom_battmgr.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/power/supply/qcom_battmgr.c b/drivers/power/supply/qcom_battmgr.c
index 80572ee945b4..644b0f5b192e 100644
--- a/drivers/power/supply/qcom_battmgr.c
+++ b/drivers/power/supply/qcom_battmgr.c
@@ -818,7 +818,7 @@ static const enum power_supply_property sc8280xp_bat_props[] = {
 };
 
 static const struct power_supply_desc sc8280xp_bat_psy_desc = {
-	.name = "qcom-battmgr-bat",
+	.name = "battery",
 	.type = POWER_SUPPLY_TYPE_BATTERY,
 	.properties = sc8280xp_bat_props,
 	.num_properties = ARRAY_SIZE(sc8280xp_bat_props),
@@ -887,7 +887,7 @@ static const enum power_supply_property sm8350_bat_props[] = {
 };
 
 static const struct power_supply_desc sm8350_bat_psy_desc = {
-	.name = "qcom-battmgr-bat",
+	.name = "battery",
 	.type = POWER_SUPPLY_TYPE_BATTERY,
 	.properties = sm8350_bat_props,
 	.num_properties = ARRAY_SIZE(sm8350_bat_props),
-- 
2.43.0

