diff --git a/tools/marvell/bin_hdr/base.mk b/tools/marvell/bin_hdr/base.mk
index 33ecf70..d1ee228 100755
--- a/tools/marvell/bin_hdr/base.mk
+++ b/tools/marvell/bin_hdr/base.mk
@@ -208,7 +208,7 @@ CPUOPTS  = -mthumb -mthumb-interwork -march=armv7 -mlittle-endian
 BH_ROOT_DIR  = $(TOPDIR)/tools/marvell/bin_hdr
 INCLUDE      = -I$(BH_ROOT_DIR)/src_ddr -I$(BH_ROOT_DIR)/src_phy/$(BOARD) -I$(BH_ROOT_DIR)/inc/common \
 	       -I$(BH_ROOT_DIR)/inc/ddr3_soc/$(INCNAME)  -I$(BH_ROOT_DIR)/inc/ddr3_soc/$(BOARD) -I$(BH_ROOT_DIR)/platform/sysEnv/$(BOARD) -I$(TOPDIR)/include -I$(BH_ROOT_DIR)/src_init/$(BOARD)
-HOSTCFLAGS   = -Wall $(INCLUDE)
+HOSTCFLAGS   = -fno-stack-protector -Wall $(INCLUDE)
 
 ifeq ($(BIN_HDR_DEBUG),1)
 DEBUG_FLAGS += -g -O0
@@ -223,7 +223,7 @@ DEBUG_MODE_FLAG = no
 endif
 endif
 
-CFLAGS   += -Wall $(INCLUDE) $(DEBUG_FLAGS) $(CPUOPTS) -msoft-float -mabi=aapcs
+CFLAGS   += -fno-stack-protector -Wall $(INCLUDE) $(DEBUG_FLAGS) $(CPUOPTS) -msoft-float -mabi=aapcs
 
 ifeq ($(BOARD),msys_bc2)
 CFLAGS   += -fPIE -fno-zero-initialized-in-bss -fno-unwind-tables
@@ -231,7 +231,7 @@ else
 CFLAGS   += -fdata-sections -ffunction-sections
 endif
 
-EXTRA_LD_FLAGS = -Wl,--gc-sections --entry=_start
+EXTRA_LD_FLAGS = -fno-stack-protector -Wl,--gc-sections --entry=_start
 
 ifeq ($(DDRTYPE),ddr4)
 CFLAGS += -DCONFIG_DDR4
