From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheSnowfield <17957399+TheSnowfield@users.noreply.github.com>
Date: Sat, 22 Feb 2025 01:09:54 +0000
Subject: rk3308: set pinmux for internal RGB output

Signed-off-by: TheSnowfield <17957399+TheSnowfield@users.noreply.github.com>
---
 drivers/gpu/drm/rockchip/rockchip_rgb.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/gpu/drm/rockchip/rockchip_rgb.c b/drivers/gpu/drm/rockchip/rockchip_rgb.c
index 111111111111..222222222222 100644
--- a/drivers/gpu/drm/rockchip/rockchip_rgb.c
+++ b/drivers/gpu/drm/rockchip/rockchip_rgb.c
@@ -8,6 +8,7 @@
 #include <linux/component.h>
 #include <linux/media-bus-format.h>
 #include <linux/of_graph.h>
+#include <linux/pinctrl/consumer.h>
 
 #include <drm/display/drm_dp_helper.h>
 #include <drm/drm_atomic_helper.h>
@@ -169,6 +170,8 @@ struct rockchip_rgb *rockchip_rgb_init(struct device *dev,
 		goto err_free_connector;
 	}
 
+	pinctrl_pm_select_default_state(dev);
+
 	return rgb;
 
 err_free_connector:
-- 
Armbian

