From b72f0fe031ab48a161637592b03012372dc6bcc7 Mon Sep 17 00:00:00 2001
From: Frank Wunderlich <frank-w@public-files.de>
Date: Mon, 16 Jun 2025 09:03:51 +0200
Subject: [PATCH 09/84] build.conf: change upload target

---
 build.conf | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/build.conf b/build.conf
index 7681a4e75ce5..c6521172e2da 100644
--- a/build.conf
+++ b/build.conf
@@ -1,5 +1,5 @@
 uploaduser=$USER
-uploadserver=r2
+uploadserver=r3
 uploaddir=/var/lib/tftp
 #uploaduser=root
 #uploadserver=192.168.0.11
-- 
2.30.2

