diff --git a/drivers/gpu/mali/mali/platform/mali400-pmu/mali_platform.c b/drivers/gpu/mali/mali/platform/mali400-pmu/mali_platform.c
index 54e50d5..1dc4f79 100644
--- a/drivers/gpu/mali/mali/platform/mali400-pmu/mali_platform.c
+++ b/drivers/gpu/mali/mali/platform/mali400-pmu/mali_platform.c
@@ -37,7 +37,7 @@ static struct clk *gpu_pll  = NULL;
 
 _mali_osk_errcode_t mali_platform_init(void)
 {
-	int freq = 252; /* 252 MHz */
+	int freq = 600; /* 600 MHz */
 
 	gpu_pll = clk_get(NULL, PLL_GPU_CLK);
 
