diff -uprNHd -x '*.orig' -x '*.rej' sun8i/drivers/base/firmware_class.c sun8i-rt/drivers/base/firmware_class.c
--- sun8i/drivers/base/firmware_class.c	2016-08-22 14:14:41.000000000 +0200
+++ sun8i-rt/drivers/base/firmware_class.c	2016-08-22 18:58:39.756320362 +0200
@@ -27,6 +27,7 @@
 #include <linux/pm.h>
 #include <linux/suspend.h>
 #include <linux/syscore_ops.h>
+#include <linux/genhd.h>
 
 #include <generated/utsrelease.h>
 
