diff -Nur a/drivers/net/wireless/bcmdhd/Kconfig b/drivers/net/wireless/bcmdhd/Kconfig
--- a/drivers/net/wireless/bcmdhd/Kconfig	2016-09-30 19:30:00.000000000 +0200
+++ b/drivers/net/wireless/bcmdhd/Kconfig	2016-10-02 10:52:10.220177127 +0200
@@ -26,6 +26,12 @@
 	help
 	  Enables WEXT support
 
+config BCMDHD_DISABLE_WOWLAN
+	bool "Disable Wake on WLAN"
+	depends on BCMDHD
+	help
+	  Disables Wake on WLAN
+
 choice
 	prompt "Enable Chip Interface"
 	depends on BCMDHD
