diff --git a/arch/arm/boot/dts/actduino_bubble_gum_sdboot_linux.dts b/arch/arm/boot/dts/actduino_bubble_gum_sdboot_linux.dts
index 30671c9..198712b 100644
--- a/arch/arm/boot/dts/actduino_bubble_gum_sdboot_linux.dts
+++ b/arch/arm/boot/dts/actduino_bubble_gum_sdboot_linux.dts
@@ -340,6 +340,20 @@
 			};
 		};				
 
+		spi0_state_default: spi0_default {
+			spi0_mfp {
+				actions,groups = "mfp1_2_0", "mfp1_4_3";
+				actions,function = "spi0";
+			};
+		};	
+
+		spi2_state_default: spi2_default {
+			spi2_mfp {
+				actions,groups = "mfp3_2";
+				actions,function = "spi2";
+			};
+		};		
+		
 		pwm0_state_default: pwm0_default {
 			pwm0_mfp {
 				actions,groups = "mfp1_31_29_ks_in2";
@@ -543,6 +557,7 @@
                                 pinctrl-0 = <&i2s0_state_default>;				
 				earphone_detect_gpios = <&gpio 35 0>;/*GPIOB4 0: high,1: low hr280ac GPIOA21 0*/
 				speaker_en_gpios = <&gpio 126 0>; /*hr280ac GPIOD(30) 0: high, 1: low*/
+				i2s_switch_gpio = <&gpio 33 0>; /* GPIOB1 1: high, route to 40Pin port, 0: low, route to audiojack */
 				earphone_output_mode = <0>; /*0: direct mode, 1: undirect mode*/
 				mic_num = <1>; /*just suport 1, 2*/
 				mic0_gain = < 0x7 0x7 >; /*<Lchannel,Rchannel> [0x0~0xF]*/
@@ -567,7 +582,7 @@
 				bl_off_voltage_diff = <400>;                      /*mV,different voltage between SYSPOWR and battery,backlight is off*/
 				support_adaptor_type = <3>;                       /*1: DCIN  2: USB  3:DCIN+USB*/
 				boot_cap_threshold = <3>;                         /*3%,the lowest capacity permitted to boot*/ 
-				ext_dcdc_exist = <1>;                             /*1:exist external dcdc  0:do not exist external dcdc*/
+				ext_dcdc_exist = <0>;                             /*1:exist external dcdc  0:do not exist external dcdc*/
 				pwms = <&pwm 3 2666 1>;                           /* 3--pwm3   2666ns--1/(24M/64)*1000000000-1   1- -active low*/
 				pwm-names = "ext_dcdc_pwm_ctl";                   /* pwm name */
 				ext_charger_exist = <0>;                          /* 1:exist external charger  0:do not exist external charger*/
@@ -874,6 +889,8 @@
 	};
 
 	spi@b0208000 {
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi2_state_default>;
 		status = "okay";
 		m25p80@0 {
 			#address-cells = <1>;
@@ -917,7 +934,7 @@
 
 		phy_addr = <3>;
 		address-bits = <48>;
-		//random-mac-address = "okay";		/*okay:     disable:*/
+		random-mac-address = "disabled";		/*okay:     disable:*/
 		//local-mac-address = [ 00 18 fe 66 66 66 ]; /* if random address is enable,the local address will be unuseful*/
 		max-frame-size = <1518>;
 
@@ -984,13 +1001,11 @@
 	usb@b0600000 {
 		vbus_otg_en_gpio = <&gpio 36 1>; /*GPIO B4*/
 		port0_host_plug_detect = <1>;
-		status = "okay";
 	};
 	
 	usb@b0700000 {
 		vbus_otg_en_gpio = <&gpio 36 1>; /*GPIO B4*/
 		port1_host_plug_detect = <1>;
-		status = "okay";
 	};
 
 	usb@b0400000 {
@@ -1062,83 +1077,37 @@
 			soc5140 = <1>;
     	};
     };
-
 	
-	framebuffer {
-		compatible = "actions,framebuffer";
-
-		/*never creat platform device*/
-		status = "disabled";
-		};
-
-		ovl_boot {
-			usage = "boot";
-			disp_dev = <&lcd0>;
-		};
-
-		ovl_extra {
-			usage = "extra";
-		};
 		
 	lcd0: lcd0@b02a0000 {
 		status = "disabled";
 	};
 
-	dsi: dsi@b0220000 {
-		pinctrl-names = "default";
-		pinctrl-0 = <&dsi_state_default>;
-		dsi_power_gpios = <&gpio 126 1>;    /*hr280ac GPIOD30*/
-		//dsi_reset_gpios = <&gpio 119 0>;    /*S500 GPIOD19*/	
-		port_type = "dsi";	 
-		data_width = <1>;
-		videomode-0 = <&dsi_mode>;
-		hw-set = <&dsi_reg>;
-
-		dsi_reg:val{													
-				dsi_ctrl =< 0x1340>;
-				dsi_size = <0x3200000>;
-				dsi_color = <0xff0000>;
-				dsi_rgbht0 = <0x7878>;
-				dsi_rgbht1 = <0x3e8>;
-				dsi_rgbvt0 = <0x1f0c32c>;
-				dsi_rgbvt1 =<0xa>;
-				dsi_pack_cfg = <0x0>;
-				dsi_pack_header =<0xf00>;
-				dsi_vedio_cfg = <0x30c>;
-				dsi_phy_t0 = <0xa5a>;
-				dsi_phy_t1 =<0x1b12>;
-				dsi_phy_t2 = <0x2f05>;
-				dsi_phy_ctrl = <0xfc94>;
-				dsi_pin_map = <0x688>;
-				cmu_dsipll_clk = <0x12601>;
-
-
-		};
-		
-		dsi_mode:mode@800x1280p60 {
-			refresh = <60>;
-			xres = <1280>;
-			yres = <800>;
-			/*in pico second, 0.000 000 000 001s*/
-			pixclock = <15000>;
-			left_margin = <10>;
-			right_margin = <10>;
-			upper_margin = <3>;
-			lower_margin = <3>;
-			hsync_len = <30>;
-			vsync_len = <6>;
-			/*0: FB_VMODE_NONINTERLACED, 1:FB_VMODE_INTERLACED*/
-			vmode = <0>;
-		};
+	dsi: dsi@b0220000 { 
+		status = "disabled";
 	};
 
 	hdmi:hdmi@b02c0000{
 		hdcp_onoff = <0>;
 		channel_invert = <0>;
 		bit_invert = <0>;
+		hotplugable = <0>;
+		bootable = <1>; 
+		bootrotate = <0>;   
+		default_resolution = "auto";
+	};
+
+	cvbs:cvbs@b02b0000{  
+		default_mode = "PAL";
 		bootable = <1>; 
-		bootrotate = <3>;   
-		default_resolution = "1080P60HZ";
+		hotplugable = <0>;
+		bootrotate = <0>;
+		status="okay";
+	};
+	
+	framebuffer {
+		compatible = "actions,framebuffer";  
+		def_display  = "hdmi";
 	};
 
 	pwm: pwm@b01b0050 {
@@ -1165,16 +1134,7 @@
 	};
 
 	backlight {
-		/*1.pwm num; 2. period in ns; */
-		/*3.plarity, 0: high active, 1: low active*/
-		backlight_en_gpios = <&gpio 128 0>; /*bubble gum GPIOE0*/ 
-		pwms = <&pwm 0 50000 1>;
-		total_steps = <1024>;
-		min_brightness = <250>;
-		max_brightness = <1000>;
-		dft_brightness = <600>;
-		delay_bf_pwm  = <200>;
-		delay_af_pwm  = <10>;	
+		status = "disabled";	
 	};
 	
 	
