diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index 878fc14..90329ab 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -222,7 +222,11 @@ dtb-$(CONFIG_ARCH_OWL) += lemaker_guitar_bba.dtb \
 		lemaker_guitar_bbb.dtb \
 		lemaker_guitar_bbb_plus.dtb \
 		lemaker_guitar_bbc.dtb \
-		lemaker_guitar_bbd.dtb
+		lemaker_guitar_bbd.dtb \
+		actduino_bubble_gum.dtb \
+		actduino_bubble_gum_sdboot.dtb \
+		actduino_bubble_gum_linux.dtb \
+		actduino_bubble_gum_sdboot_linux.dtb
 
 ifeq ($(CONFIG_ARCH_OWL),y)
 	DTC_FLAGS = -R -4 -p 0x1000
