From 8216f17965de7bc7ced7092aab0e2bfe416838a4 Mon Sep 17 00:00:00 2001
From: Stephen <stephen@vamrs.com>
Date: Sat, 22 Jun 2019 10:41:45 +0800
Subject: [PATCH 60/97] drivers: add xtx spi nor flash xt25f32b

Signed-off-by: Stephen <stephen@vamrs.com>
---
 drivers/mtd/spi-nor/spi-nor.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/mtd/spi-nor/spi-nor.c b/drivers/mtd/spi-nor/spi-nor.c
index 64d6f053c2a5..287a09052c14 100644
--- a/drivers/mtd/spi-nor/spi-nor.c
+++ b/drivers/mtd/spi-nor/spi-nor.c
@@ -845,6 +845,9 @@ static const struct flash_info spi_nor_ids[] = {
 	{ "cat25c09", CAT25_INFO( 128, 8, 32, 2, SPI_NOR_NO_ERASE | SPI_NOR_NO_FR) },
 	{ "cat25c17", CAT25_INFO( 256, 8, 32, 2, SPI_NOR_NO_ERASE | SPI_NOR_NO_FR) },
 	{ "cat25128", CAT25_INFO(2048, 8, 64, 2, SPI_NOR_NO_ERASE | SPI_NOR_NO_FR) },
+
+	/* XTX SPI Nor Flash */
+	{ "xt25f32b", INFO(0x0b4016, 0, 64 * 1024, 64, SECT_4K) },
 	{ },
 };
 
-- 
2.25.1

