--- a/arch/arm/boot/dts/meson8b.dtsi
+++ b/arch/arm/boot/dts/meson8b.dtsi
@@ -391,28 +391,6 @@
 			nvmem-cells = <&cvbs_trimming>;
 			nvmem-cell-names = "cvbs_trimming";
 
-			assigned-clocks = <&clkc CLKID_HDMI_PLL_DCO>,
-					  <&clkc CLKID_HDMI_PLL_LVDS_OUT>,
-					  <&clkc CLKID_HDMI_PLL_HDMI_OUT>,
-					  <&clkc CLKID_PLL_VID>,
-					  <&clkc CLKID_VID_PLL_FINAL_DIV>,
-					  <&clkc CLKID_HDMI_TX_PIXEL>,
-					  <&clkc CLKID_CTS_ENCP>,
-					  <&clkc CLKID_CTS_ENCI>,
-					  <&clkc CLKID_CTS_ENCT>,
-					  <&clkc CLKID_CTS_ENCL>,
-					  <&clkc CLKID_CTS_VDAC0>;
-			assigned-clock-rates = <1296000000>, /* CLKID_HDMI_PLL_DCO */
-					       <1296000000>, /* CLKID_HDMI_PLL_LVDS_OUT */
-						<324000000>, /* CLKID_HDMI_PLL_HDMI_OUT */
-						<216000000>, /* CLKID_PLL_VID */
-						 <54000000>, /* CLKID_VID_PLL_FINAL_DIV */
-						 <27000000>, /* CLKID_HDMI_TX_PIXEL */
-						 <54000000>, /* CLKID_CTS_ENCP */
-						 <27000000>, /* CLKID_CTS_ENCI */
-						 <54000000>, /* CLKID_CTS_ENCT */
-						 <54000000>, /* CLKID_CTS_ENCL */
-						 <27000000>; /* CLKID_CTS_VDAC0 */
 
 			#address-cells = <1>;
 			#size-cells = <0>;

