diff --git a/arch/arm/dts/socfpga_agilex5_socdk-u-boot.dtsi b/arch/arm/dts/socfpga_agilex5_socdk-u-boot.dtsi
--- a/arch/arm/dts/socfpga_agilex5_socdk-u-boot.dtsi
+++ b/arch/arm/dts/socfpga_agilex5_socdk-u-boot.dtsi
@@ -67,7 +67,7 @@
 
 	chosen {
 		stdout-path = "serial0:115200n8";
-		u-boot,spl-boot-order = &mmc,&flash0,&nand,"/memory";
+		u-boot,spl-boot-order = &mmc;
 	};
 };
 
@@ -158,7 +158,7 @@
 };
 
 &nand {
-	bootph-all;
+	status = "disabled";
 };
 
 &timer0 {
diff --git a/arch/arm/dts/socfpga_agilex5_socdk.dts b/arch/arm/dts/socfpga_agilex5_socdk.dts
--- a/arch/arm/dts/socfpga_agilex5_socdk.dts
+++ b/arch/arm/dts/socfpga_agilex5_socdk.dts
@@ -75,8 +75,8 @@
 };
 
 &mmc {
-	sd-uhs-sdr104;
-	sdhci-caps-mask = <0x00002000 0x0000ff00>;
+	sdhci-caps-mask = <0x00002007 0x0000ff00>;
+	no-1-8-v;
 };
 
 &uart0 {
