From f690153325e15ce41c2053d9278138b5255d96d9 Mon Sep 17 00:00:00 2001
From: Frank Wunderlich <frank-w@public-files.de>
Date: Sun, 26 Oct 2025 15:10:01 +0100
Subject: [PATCH 072/116] arm64: dts: r4pro: add eth-mux

---
 .../mt7988a-bananapi-bpi-r4-pro-8x.dts        | 31 +++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/arch/arm64/boot/dts/mediatek/mt7988a-bananapi-bpi-r4-pro-8x.dts b/arch/arm64/boot/dts/mediatek/mt7988a-bananapi-bpi-r4-pro-8x.dts
index 2e95eebb4bf6..656bbe2d3808 100644
--- a/arch/arm64/boot/dts/mediatek/mt7988a-bananapi-bpi-r4-pro-8x.dts
+++ b/arch/arm64/boot/dts/mediatek/mt7988a-bananapi-bpi-r4-pro-8x.dts
@@ -15,6 +15,37 @@ / {
 		     "mediatek,mt7988a";
 };
 
+&eth {
+	mux: mux-bus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		mux1: ethernet-mux@1 {
+			compatible = "mediatek,eth-mux";
+			reg = <1>; //map to gmac1
+			chan-sel-gpios = <&pio 3 GPIO_ACTIVE_HIGH>;
+			mod-def0-gpios = <&pio 1 GPIO_ACTIVE_LOW>; //also moddef0 of sfp2
+			sfp-present-channel = <1>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			channel_10: channel@0 {
+				reg = <0>;
+				phy-mode = "10gbase-r";
+				phy-connection-type = "10gbase-r";
+				phy-handle = <&phy28>;
+			};
+
+			channel_11: channel@1 {
+				reg = <1>;
+				phy-mode = "10gbase-r";
+				phy-connection-type = "10gbase-r";
+				managed = "in-band-status";
+				sfp = <&sfp2>;
+			};
+		};
+	};
+};
+
 &gmac1 {
 	phy-mode = "usxgmii";
 	phy-connection-type = "usxgmii";
-- 
2.39.5

