From f5cc80451af750f45c135b800f9c290e0174c4d7 Mon Sep 17 00:00:00 2001
From: Frank Wunderlich <frank-w@public-files.de>
Date: Wed, 3 Sep 2025 19:31:28 +0200
Subject: [PATCH 086/116] cpufreq: mediatek: Add support for MT7987

---
 drivers/cpufreq/mediatek-cpufreq.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/cpufreq/mediatek-cpufreq.c b/drivers/cpufreq/mediatek-cpufreq.c
index 79447a89bec8..fa7271e29bcd 100644
--- a/drivers/cpufreq/mediatek-cpufreq.c
+++ b/drivers/cpufreq/mediatek-cpufreq.c
@@ -712,6 +712,12 @@ static const struct mtk_cpufreq_platform_data mt7623_platform_data = {
 	.ccifreq_supported = false,
 };
 
+static const struct mtk_cpufreq_platform_data mt7987_platform_data = {
+	.proc_max_volt = 1023000,
+	.ccifreq_supported = false,
+	.proc_fixed_volt = true,
+};
+
 static const struct mtk_cpufreq_platform_data mt7988_platform_data = {
 	.min_volt_shift = 100000,
 	.max_volt_shift = 200000,
@@ -754,6 +760,7 @@ static const struct of_device_id mtk_cpufreq_machines[] __initconst __maybe_unus
 	{ .compatible = "mediatek,mt2712", .data = &mt2701_platform_data },
 	{ .compatible = "mediatek,mt7622", .data = &mt7622_platform_data },
 	{ .compatible = "mediatek,mt7623", .data = &mt7623_platform_data },
+	{ .compatible = "mediatek,mt7987", .data = &mt7987_platform_data },
 	{ .compatible = "mediatek,mt7988a", .data = &mt7988_platform_data },
 	{ .compatible = "mediatek,mt8167", .data = &mt8516_platform_data },
 	{ .compatible = "mediatek,mt817x", .data = &mt2701_platform_data },
-- 
2.39.5

