From 1621d837f79133302da67cc489bd505701fa55d9 Mon Sep 17 00:00:00 2001
From: Frank Wunderlich <frank-w@public-files.de>
Date: Sun, 9 Nov 2025 23:11:19 +0100
Subject: [PATCH 108/116] its: r4: enable bpi-r4 lite

---
 bpi-r4.its | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/bpi-r4.its b/bpi-r4.its
index 6f4bfce9631a..3051eccc975f 100644
--- a/bpi-r4.its
+++ b/bpi-r4.its
@@ -62,7 +62,7 @@
 				algo = "sha1";
 			};
 		};
-		/*fdt-base-lite {
+		fdt-base-lite {
 			description = "Flattened Device Tree blob BPI-R4 Lite";
 			data = /incbin/("./arch/arm64/boot/dts/mediatek/mt7987a-bananapi-bpi-r4-lite.dtb");
 			type = "flat_dt";
@@ -72,7 +72,7 @@
 			hash-1 {
 				algo = "sha1";
 			};
-		};*/
+		};
 		fdt-ov-sd {
 			description = "Flattened Device Tree blob";
 			data = /incbin/("./arch/arm64/boot/dts/mediatek/mt7988a-bananapi-bpi-r4-sd.dtbo");
@@ -83,7 +83,7 @@
 			data = /incbin/("./arch/arm64/boot/dts/mediatek/mt7988a-bananapi-bpi-r4-emmc.dtbo");
 			type = "flat_dt";
 		};
-/*		fdt-ov-lite-sd {
+		fdt-ov-lite-sd {
 			description = "Flattened Device Tree blob";
 			data = /incbin/("./arch/arm64/boot/dts/mediatek/mt7987a-bananapi-bpi-r4-lite-sd.dtbo");
 			type = "flat_dt";
@@ -112,7 +112,7 @@
 			description = "DT overlay for R4Lite 2 pcie a 1 Lanes";
 			data = /incbin/("./arch/arm64/boot/dts/mediatek/mt7987a-bananapi-bpi-r4-lite-2pcie-1L.dtbo");
 			type = "flat_dt";
-		};*/
+		};
 		fdt-ov-pro-cn15 {
 			description = "DT overlay for R4Pro CN15";
 			data = /incbin/("./arch/arm64/boot/dts/mediatek/mt7988a-bananapi-bpi-r4-pro-cn15.dtbo");
@@ -192,7 +192,7 @@
 				algo = "sha1";
 			};
 		};
-		/*conf-base-r4lite {
+		conf-base-r4lite {
 			description = "Boot Linux kernel with base FDT blob for BPI-R4 Lite";
 			kernel = "kernel-1";
 			fdt = "fdt-base-lite";
@@ -200,7 +200,7 @@
 			hash-1 {
 				algo = "sha1";
 			};
-		};*/
+		};
 		conf-sd {
 			description = "Boot Linux kernel with SD FDT blob";
 			kernel = "kernel-1";
@@ -235,7 +235,7 @@
 			description = "EMMC FDT overlay blob";
 			fdt = "fdt-ov-pro-emmc";
 		};
-		/*ov-lite-sd {
+		ov-lite-sd {
 			description = "SD FDT overlay blob for R4 Lite";
 			fdt = "fdt-ov-lite-sd";
 		};
@@ -258,7 +258,7 @@
 		ov-lite-2pcie1l {
 			description = "2 PCIe 1Lane FDT overlay blob for R4 Lite";
 			fdt = "fdt-ov-lite-2pcie";
-		};*/
+		};
 		ov-pro-cn15 {
 			description = "CN15 FDT overlay blob";
 			fdt = "fdt-ov-pro-cn15";
-- 
2.39.5

