diff --git a/arch/arm/boot/dts/rk3288-miniarm.dts b/arch/arm/boot/dts/rk3288-miniarm.dts
index 348e9be..3896506 100644
--- a/arch/arm/boot/dts/rk3288-miniarm.dts
+++ b/arch/arm/boot/dts/rk3288-miniarm.dts
@@ -194,6 +194,16 @@
 	status = "ok";
 };
 
+&mipi_dsi {
+	status = "okay";
+	mipi_panel: mipi-panel {
+		compatible ="asus,tc358762";
+		reg = <0>;
+		status = "okay";
+	};
+};
+
+
 &hdmi {
 	ddc-i2c-bus = <&i2c5>;
 	status = "okay";
@@ -400,9 +410,15 @@
 &i2c2 {
 	status = "okay";
 
+	afc0:af-controller@0 {
+		status = "okay";
+		compatible = "silicon touch,vm149C-v4l2-i2c-subdev";
+		reg = < 0x0c >;
+	};
+
 	eeprom:m24c08@50 {
-           compatible = "at,24c08";
-           reg = <0x50>;
+        compatible = "at,24c08";
+        reg = <0x50>;
     };
 };
 
@@ -626,6 +642,7 @@
 };
 
 &usb_host0_ehci {
+	no-relinquish-port;
 	status = "okay";
 };