From 835401d628f5e5850900be90a061271fb817a73e Mon Sep 17 00:00:00 2001
From: Philippe Simons <simons.philippe@gmail.com>
Date: Mon, 19 Jan 2026 12:53:14 +0100
Subject: [PATCH] power:supply: rename qcom battmgr sysfs

---
 drivers/power/supply/qcom_battmgr.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/power/supply/qcom_battmgr.c b/drivers/power/supply/qcom_battmgr.c
index 3c2837ef3461..157c2d965931 100644
--- a/drivers/power/supply/qcom_battmgr.c
+++ b/drivers/power/supply/qcom_battmgr.c
@@ -931,7 +931,7 @@ static const enum power_supply_property sm8550_bat_props[] = {
 };
 
 static const struct power_supply_desc sm8550_bat_psy_desc = {
-	.name = "qcom-battmgr-bat",
+	.name = "battery",
 	.type = POWER_SUPPLY_TYPE_BATTERY,
 	.properties = sm8550_bat_props,
 	.num_properties = ARRAY_SIZE(sm8550_bat_props),
-- 
2.52.0

