diff --git a/arch/arm64/boot/dts/nexell/Makefile b/arch/arm64/boot/dts/nexell/Makefile
index 65c2ac6..e0be747
--- a/arch/arm64/boot/dts/nexell/Makefile
+++ b/arch/arm64/boot/dts/nexell/Makefile
@@ -1,14 +1,6 @@
-dtb-$(CONFIG_ARCH_S5P6818) += s5p6818-artik710-raptor-rev00.dtb
-dtb-$(CONFIG_ARCH_S5P6818) += s5p6818-artik710-raptor-rev01.dtb
-dtb-$(CONFIG_ARCH_S5P6818) += s5p6818-artik710-raptor-rev03.dtb
-dtb-$(CONFIG_ARCH_S5P6818) += s5p6818-artik710-explorer.dtb
-dtb-$(CONFIG_ARCH_S5P6818) += s5p6818-drone.dtb
-dtb-$(CONFIG_ARCH_S5P6818) += s5p6818-avn-ref-rev01.dtb
-dtb-$(CONFIG_ARCH_S5P6818) += s5p6818-kick-st-rev00.dtb
 dtb-$(CONFIG_ARCH_S5P6818) += s5p6818-nanopi3-rev01.dtb
 dtb-$(CONFIG_ARCH_S5P6818) += s5p6818-nanopi3-rev05.dtb
 dtb-$(CONFIG_ARCH_S5P6818) += s5p6818-nanopi3-rev07.dtb
-dtb-$(CONFIG_ARCH_NXP5540) += nxp5540-evb.dtb

 always		:= $(dtb-y)
 subdir-y	:= $(dts-dirs)
