From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jianfeng Liu <Jianfeng.Liu@cixtech.com>
Date: Wed, 18 Mar 2026 14:29:38 +0800
Subject: mailbox: add acpi support to cix-mailbox driver

Signed-off-by: Jianfeng Liu <Jianfeng.Liu@cixtech.com>
---
 drivers/mailbox/cix-mailbox.c | 44 +++++++---
 1 file changed, 34 insertions(+), 10 deletions(-)

diff --git a/drivers/mailbox/cix-mailbox.c b/drivers/mailbox/cix-mailbox.c
index 111111111111..222222222222 100644
--- a/drivers/mailbox/cix-mailbox.c
+++ b/drivers/mailbox/cix-mailbox.c
@@ -3,6 +3,7 @@
  * Copyright 2025 Cix Technology Group Co., Ltd.
  */
 
+#include <linux/acpi.h>
 #include <linux/device.h>
 #include <linux/err.h>
 #include <linux/io.h>
@@ -560,6 +561,7 @@ static int cix_mbox_probe(struct platform_device *pdev)
 	struct cix_mbox_priv *priv;
 	struct resource *res;
 	const char *dir_str;
+	u32 dir;
 	int ret;
 
 	priv = devm_kzalloc(dev, sizeof(*priv), GFP_KERNEL);
@@ -587,18 +589,33 @@ static int cix_mbox_probe(struct platform_device *pdev)
 	if (priv->irq < 0)
 		return priv->irq;
 
-	if (device_property_read_string(dev, "cix,mbox-dir", &dir_str)) {
-		dev_err(priv->dev, "cix,mbox_dir property not found\n");
-		return -EINVAL;
+	if (ACPI_COMPANION(dev)) {
+		if (device_property_read_u32(dev, "cix,mbox_dir", &dir)) {
+			dev_err(priv->dev, "cix,mbox_dir property not found\n");
+			return -EINVAL;
+		}
+	} else {
+		if (device_property_read_string(dev, "cix,mbox-dir", &dir_str)) {
+			dev_err(priv->dev, "cix,mbox_dir property not found\n");
+			return -EINVAL;
+		}
 	}
 
-	if (!strcmp(dir_str, "tx"))
-		priv->dir = 0;
-	else if (!strcmp(dir_str, "rx"))
-		priv->dir = 1;
-	else {
-		dev_err(priv->dev, "cix,mbox_dir=%s is not expected\n", dir_str);
-		return -EINVAL;
+	if (ACPI_COMPANION(dev)) {
+		if ((dir != CIX_MBOX_TX) && (dir != CIX_MBOX_RX)) {
+			dev_err(priv->dev, "Dir value is not expected! dir %d\n", dir);
+			return -EINVAL;
+		}
+		priv->dir = (int)dir;
+	} else {
+		if (!strcmp(dir_str, "tx"))
+			priv->dir = 0;
+		else if (!strcmp(dir_str, "rx"))
+			priv->dir = 1;
+		else {
+			dev_err(priv->dev, "cix,mbox_dir=%s is not expected\n", dir_str);
+			return -EINVAL;
+		}
 	}
 
 	cix_mbox_init(priv);
@@ -624,11 +641,18 @@ static const struct of_device_id cix_mbox_dt_ids[] = {
 };
 MODULE_DEVICE_TABLE(of, cix_mbox_dt_ids);
 
+static const struct acpi_device_id cix_mbox_acpi_match[] = {
+        { "CIXHA001", 0 }, /* sky1 mailbox */
+        { },
+};
+MODULE_DEVICE_TABLE(acpi, cix_mbox_acpi_match);
+
 static struct platform_driver cix_mbox_driver = {
 	.probe = cix_mbox_probe,
 	.driver = {
 		.name = "cix_mbox",
 		.of_match_table = cix_mbox_dt_ids,
+		.acpi_match_table = ACPI_PTR(cix_mbox_acpi_match),
 	},
 };
 
-- 
Armbian

