From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jianfeng Liu <Jianfeng.Liu@cixtech.com>
Date: Wed, 18 Mar 2026 14:29:38 +0800
Subject: clk: clk-scmi: register clkdev for acpi

Signed-off-by: Jianfeng Liu <Jianfeng.Liu@cixtech.com>
---
 drivers/clk/clk-scmi.c | 23 ++++++++--
 1 file changed, 19 insertions(+), 4 deletions(-)

diff --git a/drivers/clk/clk-scmi.c b/drivers/clk/clk-scmi.c
index 111111111111..222222222222 100644
--- a/drivers/clk/clk-scmi.c
+++ b/drivers/clk/clk-scmi.c
@@ -5,8 +5,10 @@
  * Copyright (C) 2018-2024 ARM Ltd.
  */
 
+#include <linux/acpi.h>
 #include <linux/bits.h>
 #include <linux/clk-provider.h>
+#include <linux/clkdev.h>
 #include <linux/device.h>
 #include <linux/err.h>
 #include <linux/of.h>
@@ -404,7 +406,7 @@ static int scmi_clocks_probe(struct scmi_device *sdev)
 	struct clk_hw **hws;
 	struct clk_hw_onecell_data *clk_data;
 	struct device *dev = &sdev->dev;
-	struct device_node *np = dev->of_node;
+	struct fwnode_handle *fwnode = dev->fwnode;
 	const struct scmi_handle *handle = sdev->handle;
 	struct scmi_protocol_handle *ph;
 	const struct clk_ops *scmi_clk_ops_db[SCMI_MAX_CLK_OPS] = {};
@@ -420,7 +422,7 @@ static int scmi_clocks_probe(struct scmi_device *sdev)
 
 	count = scmi_proto_clk_ops->count_get(ph);
 	if (count < 0) {
-		dev_err(dev, "%pOFn: invalid clock output count\n", np);
+		dev_err(dev, "%s: invalid clock output count\n", fwnode_get_name(fwnode));
 		return -EINVAL;
 	}
 
@@ -492,11 +494,24 @@ static int scmi_clocks_probe(struct scmi_device *sdev)
 			dev_dbg(dev, "Registered clock:%s%s\n",
 				sclk->info->name,
 				scmi_ops->enable ? " (atomic ops)" : "");
+			if (ACPI_COMPANION(dev)) {
+				char con_id[20];
+				snprintf(con_id, sizeof(con_id), "scmi-clk-%d", idx);
+				err = devm_clk_hw_register_clkdev(dev, hws[idx], con_id, NULL);
+				if (err)
+				{
+					dev_warn(dev, "Failed to register clkdev for clock name %s index %d\n", sclk->info->name, idx);
+					break;
+				}
+			}
 		}
 	}
 
-	return devm_of_clk_add_hw_provider(dev, of_clk_hw_onecell_get,
-					   clk_data);
+	if (ACPI_COMPANION(dev))
+		return err;
+	else
+		return devm_of_clk_add_hw_provider(dev, of_clk_hw_onecell_get,
+						   clk_data);
 }
 
 static const struct scmi_device_id scmi_id_table[] = {
-- 
Cixtech

