From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jianfeng Liu <Jianfeng.Liu@cixtech.com>
Date: Wed, 18 Mar 2026 14:29:41 +0800
Subject: mfd: syscon: add acpi support for cix soc

Signed-off-by: Jianfeng Liu <Jianfeng.Liu@cixtech.com>
---
 drivers/mfd/syscon.c | 66 ++++++++++
 1 file changed, 66 insertions(+)

diff --git a/drivers/mfd/syscon.c b/drivers/mfd/syscon.c
index 111111111111..222222222222 100644
--- a/drivers/mfd/syscon.c
+++ b/drivers/mfd/syscon.c
@@ -8,6 +8,7 @@
  * Author: Dong Aisheng <dong.aisheng@linaro.org>
  */
 
+#include <linux/acpi.h>
 #include <linux/cleanup.h>
 #include <linux/clk.h>
 #include <linux/err.h>
@@ -16,6 +17,7 @@
 #include <linux/mutex.h>
 #include <linux/of.h>
 #include <linux/of_address.h>
+#include <linux/platform_device.h>
 #include <linux/regmap.h>
 #include <linux/reset.h>
 #include <linux/mfd/syscon.h>
@@ -360,3 +362,67 @@ struct regmap *syscon_regmap_lookup_by_phandle_optional(struct device_node *np,
 	return regmap;
 }
 EXPORT_SYMBOL_GPL(syscon_regmap_lookup_by_phandle_optional);
+
+#ifdef CONFIG_ACPI
+struct syscon_platform_data {
+	const char *label;
+};
+
+static int syscon_probe(struct platform_device *pdev)
+{
+	struct device *dev = &pdev->dev;
+	struct syscon_platform_data *pdata = dev_get_platdata(dev);
+	struct syscon *syscon;
+	struct regmap_config syscon_config = syscon_regmap_config;
+	struct resource *res;
+	void __iomem *base;
+
+	syscon = devm_kzalloc(dev, sizeof(*syscon), GFP_KERNEL);
+	if (!syscon)
+		return -ENOMEM;
+
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+	if (!res)
+		return -ENOENT;
+
+	base = devm_ioremap(dev, res->start, resource_size(res));
+	if (!base)
+		return -ENOMEM;
+
+	syscon_config.max_register = resource_size(res) - 4;
+	if (pdata)
+		syscon_config.name = pdata->label;
+
+	syscon->regmap = devm_regmap_init_mmio(dev, base, &syscon_config);
+	if (IS_ERR(syscon->regmap)) {
+		dev_err(dev, "regmap init failed\n");
+		return PTR_ERR(syscon->regmap);
+	}
+
+	platform_set_drvdata(pdev, syscon);
+
+	dev_dbg(dev, "regmap %pR registered\n", res);
+
+	return 0;
+}
+
+static const struct acpi_device_id syscon_acpi_match[] = {
+	{ "CIXHA018", 0 },
+	{ }
+};
+MODULE_DEVICE_TABLE(acpi, syscon_acpi_match);
+
+static struct platform_driver syscon_driver = {
+	.driver = {
+		.name = "syscon",
+		.acpi_match_table = ACPI_PTR(syscon_acpi_match),
+	},
+	.probe = syscon_probe,
+};
+
+static int __init syscon_init(void)
+{
+	return platform_driver_register(&syscon_driver);
+}
+postcore_initcall(syscon_init);
+#endif
-- 
Cixtech

