diff --git a/arch/arm/dts/rk3399-rockpro64.dts b/arch/arm/dts/rk3399-rockpro64.dts
index 1f2394e0..4e8feebb 100644
--- a/arch/arm/dts/rk3399-rockpro64.dts
+++ b/arch/arm/dts/rk3399-rockpro64.dts
@@ -116,6 +116,20 @@
 		vin-supply = <&vcc5v0_sys>;
 	};
 
+	vcc3v0_sd: vcc3v0-sd {
+		u-boot,dm-spl;
+		compatible = "regulator-fixed";
+		enable-active-high;
+		gpio = <&gpio0 RK_PA1 GPIO_ACTIVE_HIGH>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&sdmmc0_pwr_h>;
+		regulator-always-on;
+		regulator-max-microvolt = <3000000>;
+		regulator-min-microvolt = <3000000>;
+		regulator-name = "vcc3v0_sd";
+		vin-supply = <&vcc3v3_sys>;
+	};
+
 	/* Actually 3 regulators (host0, 1, 2) controlled by the same gpio */
 	vcc5v0_host: vcc5v0-host-regulator {
 		compatible = "regulator-fixed";
@@ -514,6 +528,10 @@
 	status = "okay";
 };
 
+&gpio0 {
+	u-boot,dm-spl;
+};
+
 &pinctrl {
 	buttons {
 		pwrbtn: pwrbtn {
@@ -557,6 +575,14 @@
 		};
 	};
 
+	sd {
+		sdmmc0_pwr_h: sdmmc0-pwr-h {
+			u-boot,dm-spl;
+			rockchip,pins =
+				<RK_GPIO0 RK_PA1 RK_FUNC_GPIO &pcfg_pull_none>;
+		};
+	};
+
 	sdio-pwrseq {
 		wifi_enable_h: wifi-enable-h {
 			rockchip,pins = <0 RK_PB2 RK_FUNC_GPIO &pcfg_pull_none>;
@@ -590,6 +616,7 @@
 };
 
 &sdmmc {
+	vmmc-supply = <&vcc3v0_sd>;
 	bus-width = <4>;
 	cap-mmc-highspeed;
 	cap-sd-highspeed;
