diff --git a/neofetch b/neofetch
index 1e4b564..f59eca6 100755
--- a/neofetch
+++ b/neofetch
@@ -967,6 +967,10 @@ get_distro() {
                     *) distro="Red Star OS $(awk -F'[^0-9*]' '$0=$2' /etc/redstar-release)"
                 esac
 
+            elif [[ -f /etc/armbian-release ]]; then
+                       . /etc/armbian-release
+                       distro="Armbian $DISTRIBUTION_CODENAME ($VERSION)"
+
             elif [[ -f /etc/siduction-version ]]; then
                 case $distro_shorthand in
                     on|tiny) distro=Siduction ;;
@@ -5336,6 +5340,28 @@ ${c1}
 EOF
         ;;
 
+        "Armbian"*)
+            set_colors 1 1
+            read -rd '' ascii_data <<'EOF'
+   ${c1}
+
+      █ █ █ █ █ █ █ █ █ █ █
+     ███████████████████████
+   ▄▄██                   ██▄▄
+   ▄▄██    ███████████    ██▄▄
+   ▄▄██   ██         ██   ██▄▄
+   ▄▄██   ██         ██   ██▄▄
+   ▄▄██   ██         ██   ██▄▄
+   ▄▄██   █████████████   ██▄▄
+   ▄▄██   ██         ██   ██▄▄
+   ▄▄██   ██         ██   ██▄▄
+   ▄▄██   ██         ██   ██▄▄
+   ▄▄██                   ██▄▄
+     ███████████████████████
+      █ █ █ █ █ █ █ █ █ █ █
+EOF
+        ;;
+
         "alpine_small")
             set_colors 4 7
             read -rd '' ascii_data <<'EOF'
