diff --git a/cache/sources/linux-kernel-worktree/6.16__arm64__arm64/drivers/net/wireless/rtl88x2bu/include/osdep_service_linux.h b/drivers/net/wireless/rtl88x2bu/include/osdep_service_linux.h
index 46b9528..fb32ed6 100644
--- a/drivers/net/wireless/rtl88x2bu/include/osdep_service_linux.h
+++ b/drivers/net/wireless/rtl88x2bu/include/osdep_service_linux.h
@@ -358,7 +358,11 @@ static inline void timer_hdl(unsigned long cntx)
 #endif
 {
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 14, 0))
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 16, 0))
+	_timer *ptimer = timer_container_of(ptimer, in_timer, timer);
+#else
 	_timer *ptimer = from_timer(ptimer, in_timer, timer);
+#endif
 #else
 	_timer *ptimer = (_timer *)cntx;
 #endif
