From 036a79900b8645eb9948c1dd7e5e8a984ba1845e Mon Sep 17 00:00:00 2001
From: Julian Sikorski <belegdol+github@gmail.com>
Date: Sat, 29 Jul 2023 11:16:28 +0000
Subject: [PATCH] Fix building on 5.4.251+ kernel

---
 include/osdep_service_linux.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/wireless/rtl8189fs/include/osdep_service_linux.h b/drivers/net/wireless/rtl8189fs/include/osdep_service_linux.h
index e1daa04..6c6d0d8 100644
--- a/drivers/net/wireless/rtl8189fs/include/osdep_service_linux.h
+++ b/drivers/net/wireless/rtl8189fs/include/osdep_service_linux.h
@@ -219,7 +219,7 @@ typedef void *timer_hdl_context;
 	#define DMA_BIT_MASK(n) (((n) == 64) ? ~0ULL : ((1ULL<<(n))-1))
 #endif
 
-#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 15, 0))
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 15, 0) && LINUX_VERSION_CODE > KERNEL_VERSION(5, 5, 0)) || LINUX_VERSION_CODE < KERNEL_VERSION(5, 4, 251)
 /* Porting from linux kernel commits 
 48eab831ae8b9f7002a533fa4235eed63ea1f1a3 
 3f6cffb8604b537e3d7ea040d7f4368689638eaf
-- 
2.34.1

