diff --git a/plat/sun50iw1p1/sunxi_power.c b/plat/sun50iw1p1/sunxi_power.c
index 298913b..3207680
--- a/plat/sun50iw1p1/sunxi_power.c
+++ b/plat/sun50iw1p1/sunxi_power.c
@@ -282,13 +282,14 @@ static int pmic_setup(const char *dt_name)

 		NOTICE("PMIC: enabled Pinebook display\n");
 	}
-
+
 	sunxi_pmic_write(0x15, 0x1a);	/* DLDO1 = VCC3V3_HDMI voltage = 3.3V */

 	sunxi_pmic_write(0x16, 0x12);	/* DLDO2 = VCC2V5_EDP voltage = 2.5V */
 	sunxi_pmic_write(0x1c, 0xa);	/* FLDO1 = VCC1V2_EDP voltage = 1.2V */
 	sunxi_pmic_write(0x91, 0x1a);	/* GPIO0LDO voltage = 3.3V */
 	sunxi_pmic_write(0x90, 0x3);	/* Enable GPIO0LDO */
+	sunxi_pmic_write(0x30, sunxi_pmic_read(0x30) | BIT(2));
 	ret = sunxi_pmic_read(0x13);
 	/* Enable FLDO1 to power up eDP bridge */
 	ret = sunxi_pmic_write(0x13, ret | 0x4);
