diff --git a/tools/rkcommon.c b/tools/rkcommon.c
index af2a483..87ef80a 100644
--- a/tools/rkcommon.c
+++ b/tools/rkcommon.c
@@ -137,6 +137,8 @@ static struct spl_info spl_infos[] = {
 	{ "rv1126", "110B", 0x10000 - 0x1000, false, RK_HEADER_V1 },
 	{ "rk1808", "RK18", 0x200000 - 0x2000, false, RK_HEADER_V1 },
 	{ "rk356x", "RK35", 0x10000 - 0x1000, false, RK_HEADER_V2 },
+	{ "rk3566", "RK35", 0x10000 - 0x1000, false, RK_HEADER_V2 },
+	{ "rk3568", "RK35", 0x10000 - 0x1000, false, RK_HEADER_V2 },
 };
 
 /**
