diff -Naur linux-5.5-rc1-old/arch/arm64/boot/dts/rockchip/rk3328-a1.dts linux-5.5-rc1-new/arch/arm64/boot/dts/rockchip/rk3328-a1.dts
--- linux-5.5-rc1-old/arch/arm64/boot/dts/rockchip/rk3328-a1.dts	2019-12-08 23:57:55.000000000 +0100
+++ linux-5.5-rc1-new/arch/arm64/boot/dts/rockchip/rk3328-a1.dts	2019-12-12 20:53:51.626666660 +0100
@@ -49,7 +49,8 @@
 
 	ir-receiver {
 		compatible = "gpio-ir-receiver";
-		gpios = <&gpio2 RK_PA2 GPIO_ACTIVE_HIGH>;
+		gpios = <&gpio2 RK_PA2 GPIO_ACTIVE_LOW>;
+		linux,rc-map-name = "rc-beelink-gs1";
 	};
 };
 
