diff --git a/arch/arm/boot/dts/sun8i-h2-plus-bananapi-m2-zero.dts b/arch/arm/boot/dts/sun8i-h2-plus-bananapi-m2-zero.dts
index 4c6704e4c..5b15a95c0 100644
--- a/arch/arm/boot/dts/sun8i-h2-plus-bananapi-m2-zero.dts
+++ b/arch/arm/boot/dts/sun8i-h2-plus-bananapi-m2-zero.dts
@@ -20,22 +20,45 @@ / {
 	aliases {
 		serial0 = &uart0;
 		serial1 = &uart1;
+		ethernet0 = &emac;
 	};
 
 	chosen {
 		stdout-path = "serial0:115200n8";
 	};
 
+	connector {
+		compatible = "hdmi-connector";
+		type = "a";
+
+		port {
+			hdmi_con_in: endpoint {
+				remote-endpoint = <&hdmi_out_con>;
+			};
+		};
+	};
+
 	leds {
 		compatible = "gpio-leds";
 
 		pwr_led {
 			label = "bananapi-m2-zero:red:pwr";
-			gpios = <&r_pio 0 10 GPIO_ACTIVE_LOW>; /* PL10 */
-			default-state = "on";
+			gpios = <&r_pio 0 10 GPIO_ACTIVE_HIGH>; /* PL10 */
+			linux,default-trigger = "heartbeat";
 		};
 	};
 
+	rfkill_bt: rfkill_bt {
+		compatible = "rfkill-gpio";
+		pinctrl-names = "default";
+		pinctrl-0 = <&bt_pwr_pin>;
+		reset-gpios = <&pio 6 12 GPIO_ACTIVE_HIGH>; /* PG12 */
+		clocks = <&osc32k>;
+		clock-frequency = <32768>;
+		rfkill-name = "sunxi-bt";
+		rfkill-type = "bluetooth";
+	};
+
 	gpio_keys {
 		compatible = "gpio-keys";
 
@@ -70,6 +93,21 @@ wifi_pwrseq: wifi_pwrseq {
 	};
 };
 
+&pio {
+	bt_pwr_pin: bt_pwr_pin {
+		pins = "PG12";
+		function = "gpio_out";
+	};
+};
+
+
+&r_pio {
+	wifi_en_pin: wifi_en_pin {
+		pins = "PL7";
+		function = "gpio_out";
+	};
+};
+
 &cpu0 {
 	cpu-supply = <&reg_vdd_cpux>;
 };
@@ -107,6 +145,24 @@ brcmf: wifi@1 {
 	};
 };
 
+&de {
+	status = "okay";
+};
+
+&hdmi {
+	status = "okay";
+};
+
+&hdmi_out {
+	hdmi_out_con: endpoint {
+		remote-endpoint = <&hdmi_con_in>;
+	};
+};
+
+&mixer0 {
+	status = "okay";
+};
+
 &ohci0 {
 	status = "okay";
 };
@@ -151,3 +207,19 @@ &usbphy {
 	 */
 	status = "okay";
 };
+
+
+&emac {
+	status = "okay";
+	phy-handle = <&int_mii_phy>;
+	phy-mode = "mii";
+	allwinner,leds-active-low;
+};
+
+&i2s2 {
+	status = "okay";
+};
+
+&hdmi_sound {
+	status = "okay";
+};
