diff --git a/arch/arm/boot/dts/rk322x-box.dtsi b/arch/arm/boot/dts/rk322x-box.dtsi
index 04330d393..608d3f386 100644
--- a/arch/arm/boot/dts/rk322x-box.dtsi
+++ b/arch/arm/boot/dts/rk322x-box.dtsi
@@ -22,9 +22,39 @@
 
 
 	chosen {
+		stdout-path = &uart2;
 		bootargs = "earlycon=uart8250,mmio32,0x11030000";
 	};
 
+	bus_intmem@10080000 {
+                compatible = "mmio-sram";
+                reg = <0x10080000 0x9000>;
+                #address-cells = <1>;
+                #size-cells = <1>;
+                ranges = <0 0x10080000 0x9000>;
+                smp-sram@0 {
+                        compatible = "rockchip,rk3066-smp-sram";
+                        reg = <0x00 0x10>;
+                };
+                ddr_sram: ddr-sram@1000 {
+                        compatible = "rockchip,rk322x-ddr-sram";
+                        reg = <0x1000 0x8000>;
+                };
+        };
+
+	fiq-debugger {
+		compatible = "rockchip,fiq-debugger";
+		interrupts = <GIC_SPI 81 IRQ_TYPE_LEVEL_HIGH>;
+		rockchip,serial-id = <2>;
+		rockchip,signal-irq = <159>;
+		rockchip,wake-irq = <0>;
+		rockchip,irq-mode-enable = <0>; /* If enable uart uses irq instead of fiq */
+		rockchip,baudrate = <115200>; /* Only 115200 and 1500000 */
+		pinctrl-names = "default";
+		pinctrl-0 = <&uart21_xfer>;
+		status = "disabled";
+	};
+
 	reserved-memory {
 		#address-cells = <1>;
 		#size-cells = <1>;
@@ -59,7 +89,7 @@
 		compatible = "simple-audio-card";
 		simple-audio-card,format = "i2s";
 		simple-audio-card,mclk-fs = <256>;
-		simple-audio-card,name = "ROCKCHIP,RK3229";
+		simple-audio-card,name = "ANALOG";
 		simple-audio-card,cpu {
 			sound-dai = <&i2s1>;
 		};
@@ -73,7 +103,7 @@
 		compatible = "simple-audio-card";
 		simple-audio-card,format = "i2s";
 		simple-audio-card,mclk-fs = <128>;
-		simple-audio-card,name = "HDMI";
+		simple-audio-card,name = "DW-I2S-HDMI";
 		simple-audio-card,cpu {
 			sound-dai = <&i2s0>;
 		};
@@ -113,7 +143,7 @@
 	spdif-sound {
 		status = "okay";
 		compatible = "simple-audio-card";
-		simple-audio-card,name = "ROCKCHIP,SPDIF";
+		simple-audio-card,name = "SPDIF";
 		simple-audio-card,cpu {
 			sound-dai = <&spdif>;
 		};
@@ -219,14 +249,17 @@
 
 &cpu1 {
 	enable-method = "psci";
+	cpu-supply = <&vdd_arm>;
 };
 
 &cpu2 {
 	enable-method = "psci";
+	cpu-supply = <&vdd_arm>;
 };
 
 &cpu3 {
 	enable-method = "psci";
+	cpu-supply = <&vdd_arm>;
 };
 
 
@@ -237,11 +270,23 @@
 &gmac {
 	assigned-clocks = <&cru SCLK_MAC_SRC>;
 	assigned-clock-rates = <50000000>;
+
 	clock_in_out = "output";
 	phy-supply = <&vcc_phy>;
 	phy-mode = "rmii";
 	phy-is-integrated;
+
+	tx_delay = < 0x30 >;
+	rx_delay = < 0x10 >;
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&phy_pins>;
+
+	snps,reset-gpio = <&gpio2 RK_PD0 GPIO_ACTIVE_LOW>;
+	snps,reset-active-low;
+
 	status = "okay";
+
 };
 
 &gpu {
@@ -273,11 +318,13 @@
 	broken-cd;
 	bus-width = <8>;
 	cap-mmc-highspeed;
-	mmc-hs200-1_8v;
 	supports-emmc;
 	disable-wp;
 	non-removable;
 	num-slots = <1>;
+	max-frequency = <150000000>;
+	clock-frequency = <50000000>;
+	/delete-property/ clock-freq-min-max;
 	/delete-property/ default-sample-phase;
 	/delete-property/ pinctrl-names;
 	/delete-property/ pinctrl-0;
@@ -288,7 +335,7 @@
 	cap-sd-highspeed;
 	card-detect-delay = <200>;
 	disable-wp;
-	max-frequency = <50000000>;
+	max-frequency = <150000000>;
 	num-slots = <1>;
 	supports-sd;
 	pinctrl-names = "default";
@@ -433,6 +480,7 @@
 	status = "okay";
 
 	u2phy1_otg: otg-port {
+		phy-supply = <&vcc_host>;
 		status = "okay";
 	};
 
@@ -443,6 +491,7 @@
 };
 
 &usb_otg {
+	dr_mode = "host";
 	status = "okay";
 };
 
