diff --git a/include/configs/udoo_qdl.h b/include/configs/udoo_qdl.h
--- a/include/configs/udoo_qdl.h
+++ b/include/configs/udoo_qdl.h
@@ -86,7 +86,11 @@
 #define CONFIG_ENV_VARS_UBOOT_RUNTIME_CONFIG
 
 #define CONFIG_EXTRA_ENV_SETTINGS \
-	"script=uEnv.txt\0" \
+	"script=boot.scr\0" \
+	"boot_prefixes=/ /boot/\0" \
+	"boot_a_script=ext2load mmc ${mmcdev}:${mmcpart} ${loadaddr} ${prefix}${script} || fatload mmc ${mmcdev}:${mmcpart} ${loadaddr} ${prefix}${script} \0" \
+	"scan_dev_for_boot=echo Scanning mmc ; for prefix in ${boot_prefixes}; do run scan_dev_for_scripts; done\0" \
+	"scan_dev_for_scripts=if test -e ${mmcdev}:${mmcpart} ${prefix}${script}; then echo Found U-Boot script ${prefix}${script}; run boot_a_script; else echo Checking here ${prefix} ... not found; fi;\0" \
 	"image=zImage\0" \
 	"console=ttymxc1\0" \
 	"splashpos=m,m\0" \
@@ -116,9 +120,9 @@
 	"mmcargs=setenv bootargs console=${console},${baudrate} ${video} ${memory} " \
 		"root=${mmcroot}\0" \
 	"loadbootscript=" \
-		"fatload " UBOOT_DEVICE " ${mmcdev}:${mmcpart} ${loadaddr} ${script};\0" \
+	"run scan_dev_for_boot;\0" \
 	"bootscript=echo Running bootscript from mmc ...; " \
-		"env import -t ${loadaddr} ${filesize};\0" \
+		"source\0" \
 	"loadimage=fatload " UBOOT_DEVICE " ${mmcdev}:${mmcpart} ${loadaddr} ${image}\0" \
 	"loadfdt=fatload " UBOOT_DEVICE " ${mmcdev}:${mmcpart} ${fdt_addr} ${fdt_file}\0" \
 	"mmcboot=echo Booting from mmc ...; " \