diff --git a/arch/arm64/boot/dts/marvell/armada-3720-espressobin.dts b/arch/arm64/boot/dts/marvell/armada-3720-espressobin.dts
old mode 100644
new mode 100755
index bb8cb31..0e5348c
--- a/arch/arm64/boot/dts/marvell/armada-3720-espressobin.dts
+++ b/arch/arm64/boot/dts/marvell/armada-3720-espressobin.dts
@@ -116,6 +116,26 @@
 	status = "okay";
 };
 
+&spi0 {
+	status = "okay";
+
+	w25q32dw@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <104000000>;
+		m25,fast-read;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_quad_pins>;
+	};
+};
+
+&i2c0 {
+	status = "okay";
+};
+
 &mdio {
 	switch0: switch0@1 {
 		compatible = "marvell,mv88e6085";
