From 6a1c3a22addf1367b76ff1e43d4428f19d88faf7 Mon Sep 17 00:00:00 2001
From: Gunjan Gupta <viraniac@gmail.com>
Date: Tue, 5 Sep 2023 11:16:07 +0000
Subject: [PATCH] drivers: devfreq: sun8i-a33-mbus: disable autorefresh

---
 drivers/devfreq/sun8i-a33-mbus.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/devfreq/sun8i-a33-mbus.c b/drivers/devfreq/sun8i-a33-mbus.c
index 13d32213139f..81832c3ffafa 100644
--- a/drivers/devfreq/sun8i-a33-mbus.c
+++ b/drivers/devfreq/sun8i-a33-mbus.c
@@ -202,8 +202,8 @@ static int sun8i_a33_mbus_set_dram_freq(struct sun8i_a33_mbus *priv,
 	writel_relaxed(vtfcr, priv->reg_dram + DRAM_VTFCR);
 
 	/* Enable automatic self-refresh at the lowest frequency only. */
-	if (freq == priv->freq_table[0])
-		pwrctl |= DRAM_PWRCTL_SELFREF_EN;
+	// if (freq == priv->freq_table[0])
+	//	pwrctl |= DRAM_PWRCTL_SELFREF_EN;
 	writel_relaxed(pwrctl, priv->reg_dram + DRAM_PWRCTL);
 
 	sun8i_a33_mbus_restart_pmu_counters(priv);
-- 
2.34.1

