From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Patrick Yavitz <pyavitz@armbian.com>
Date: Fri, 14 Jun 2024 10:39:21 -0400
Subject: overlay support

Signed-off-by: Patrick Yavitz <pyavitz@armbian.com>
---
 configs/k1_defconfig   | 1 +
 include/configs/k1-x.h | 2 ++
 2 files changed, 3 insertions(+)

diff --git a/configs/k1_defconfig b/configs/k1_defconfig
index 111111111111..222222222222 100644
--- a/configs/k1_defconfig
+++ b/configs/k1_defconfig
@@ -272,3 +272,4 @@ CONFIG_IMAGE_SPARSE_TRANSFER_BLK_NUM=0x3000
 # CONFIG_SPL_SHA1 is not set
 # CONFIG_SPL_SHA256 is not set
 CONFIG_ZSTD=y
+CONFIG_OF_LIBFDT_OVERLAY=y
diff --git a/include/configs/k1-x.h b/include/configs/k1-x.h
index 111111111111..222222222222 100644
--- a/include/configs/k1-x.h
+++ b/include/configs/k1-x.h
@@ -141,6 +141,7 @@ struct boot_storage_op
 #define KERNEL_ADDR_R		__stringify(0x10000000)
 #define FDT_ADDR_R		__stringify(0x1F000000)
 #define RAMDISK_ADDR_R		__stringify(0x21000000)
+#define FDTOVERLAY_ADDR_R	__stringify(0x01000000)
 
 #include <config_distro_bootcmd.h>
 
@@ -172,6 +173,7 @@ struct boot_storage_op
 	"fdt_addr_r=" FDT_ADDR_R "\0" \
 	"kernel_addr_r=" KERNEL_ADDR_R "\0" \
 	"ramdisk_addr_r=" RAMDISK_ADDR_R "\0" \
+	"fdtoverlay_addr_r=" FDTOVERLAY_ADDR_R "\0" \
 	"fdtfile=spacemit/" CONFIG_DEFAULT_DEVICE_TREE ".dtb\0" \
 	"dtb_addr=" __stringify(DTB_LOAD_ADDR) "\0" \
 	"scriptaddr=0x2c100000\0" \
-- 
Armbian

