From 4e65b4371d252c546a2ed996900839e09b25a110 Mon Sep 17 00:00:00 2001
From: Patrick Yavitz <pyavitz@armbian.com>
Date: Fri, 14 Jun 2024 10:39:21 -0400
Subject: [PATCH] overlay support

Signed-off-by: Patrick Yavitz <pyavitz@armbian.com>
---
 configs/k1_defconfig   | 1 +
 include/configs/k1-x.h | 2 ++
 2 files changed, 3 insertions(+)

diff --git a/configs/k1_defconfig b/configs/k1_defconfig
index 969e38e2e7..f009ebca5c 100644
--- a/configs/k1_defconfig
+++ b/configs/k1_defconfig
@@ -270,3 +270,4 @@ CONFIG_IMAGE_SPARSE_TRANSFER_BLK_NUM=0x3000
 # CONFIG_SPL_SHA1 is not set
 # CONFIG_SPL_SHA256 is not set
 CONFIG_ZSTD=y
+CONFIG_OF_LIBFDT_OVERLAY=y
diff --git a/include/configs/k1-x.h b/include/configs/k1-x.h
index a18e5bf18a..6980b33c6b 100644
--- a/include/configs/k1-x.h
+++ b/include/configs/k1-x.h
@@ -135,6 +135,7 @@ struct boot_storage_op
 #define KERNEL_ADDR_R		__stringify(0x10000000)
 #define FDT_ADDR_R		__stringify(0x1F000000)
 #define RAMDISK_ADDR_R		__stringify(0x21000000)
+#define FDTOVERLAY_ADDR_R	__stringify(0x01000000)
 
 #include <config_distro_bootcmd.h>
 
@@ -166,6 +167,7 @@ struct boot_storage_op
 	"fdt_addr_r=" FDT_ADDR_R "\0" \
 	"kernel_addr_r=" KERNEL_ADDR_R "\0" \
 	"ramdisk_addr_r=" RAMDISK_ADDR_R "\0" \
+	"fdtoverlay_addr_r=" FDTOVERLAY_ADDR_R "\0" \
 	"fdtfile=spacemit/" CONFIG_DEFAULT_DEVICE_TREE ".dtb\0" \
 	"scriptaddr=0x2c100000\0" \
 	"pxefile_addr_r=0x0c200000\0" \
-- 
2.39.2

