diff -ruN a/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart0.dts b/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart0.dts
--- a/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart0.dts	1970-01-01 08:00:00.000000000 +0800
+++ b/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart0.dts	2024-08-01 10:43:17.000000000 +0800
@@ -0,0 +1,21 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/rockchip.h>
+#include <dt-bindings/power/rk3568-power.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {		
+	fragment@0 {
+		target = <&uart0>;
+
+		__overlay__ {
+			pinctrl-names = "default";
+			pinctrl-0 = <&uart0_xfer>;
+			status = "okay";
+		
+		};
+	};
+	
+};	
diff -ruN a/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart1.dts b/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart1.dts
--- a/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart1.dts	1970-01-01 08:00:00.000000000 +0800
+++ b/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart1.dts	2024-08-01 10:43:17.000000000 +0800
@@ -0,0 +1,21 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/rockchip.h>
+#include <dt-bindings/power/rk3568-power.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {		
+	fragment@0 {
+		target = <&uart1>;
+
+		__overlay__ {
+			pinctrl-names = "default";
+			pinctrl-0 = <&uart1m1_xfer>;
+			status = "okay";
+		
+		};
+	};
+	
+};	
diff -ruN a/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart2.dts b/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart2.dts
--- a/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart2.dts	1970-01-01 08:00:00.000000000 +0800
+++ b/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart2.dts	2024-08-01 10:43:17.000000000 +0800
@@ -0,0 +1,21 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/rockchip.h>
+#include <dt-bindings/power/rk3568-power.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {		
+	fragment@0 {
+		target = <&uart2>;
+
+		__overlay__ {
+			pinctrl-names = "default";
+			pinctrl-0 = <&uart2m0_xfer>;
+			status = "okay";
+		
+		};
+	};
+	
+};	
diff -ruN a/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart3.dts b/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart3.dts
--- a/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart3.dts	1970-01-01 08:00:00.000000000 +0800
+++ b/arch/arm64/boot/dts/rockchip/overlay/rk3566-uart3.dts	2024-08-01 10:43:17.000000000 +0800
@@ -0,0 +1,18 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/rockchip.h>
+#include <dt-bindings/power/rk3568-power.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {		
+	fragment@0 {
+		target = <&uart3>;
+
+		__overlay__ {
+			status = "okay";		
+		};
+	};
+	
+};	
