Index: drivers/spi/spi-sun4i.c
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- a/drivers/spi/spi-sun4i.c	(date 1581011681000)
+++ b/drivers/spi/spi-sun4i.c	(date 1581070363000)
@@ -389,6 +389,7 @@
 	struct spi_master *master = dev_get_drvdata(dev);
 	struct sun4i_spi *sspi = spi_master_get_devdata(master);
 	int ret;
+	u32 reg;

 	ret = clk_prepare_enable(sspi->hclk);
 	if (ret) {
@@ -401,9 +402,10 @@
 		dev_err(dev, "Couldn't enable module clock\n");
 		goto err;
 	}
+	reg = sun4i_spi_read(sspi, SUN4I_CTL_REG);

 	sun4i_spi_write(sspi, SUN4I_CTL_REG,
-			SUN4I_CTL_ENABLE | SUN4I_CTL_MASTER | SUN4I_CTL_TP);
+        reg | SUN4I_CTL_ENABLE | SUN4I_CTL_MASTER | SUN4I_CTL_TP);

 	return 0;

