diff --git a/drivers/net/wireless/Kconfig b/drivers/net/wireless/Kconfig
index 561b71c..7779c1d 100644
--- a/drivers/net/wireless/Kconfig
+++ b/drivers/net/wireless/Kconfig
@@ -31,6 +31,7 @@ config WIRELESS_WDS
 	  necessary, give up on our plan of removing it).

 source "drivers/net/wireless/admtek/Kconfig"
+source "drivers/net/wireless/rtl8189es/Kconfig"
 source "drivers/net/wireless/ath/Kconfig"
 source "drivers/net/wireless/rtl8188eu/Kconfig"
 source "drivers/net/wireless/atmel/Kconfig"
diff --git a/drivers/net/wireless/Makefile b/drivers/net/wireless/Makefile
index 614ddf4..ab14e53 100644
--- a/drivers/net/wireless/Makefile
+++ b/drivers/net/wireless/Makefile
@@ -4,6 +4,7 @@
 #

 obj-$(CONFIG_WLAN_VENDOR_ADMTEK) += admtek/
+obj-$(CONFIG_RTL8189ES) += rtl8189es/
 obj-$(CONFIG_WLAN_VENDOR_ATH) += ath/
 obj-$(CONFIG_RTL8188EU) += rtl8188eu/
 obj-$(CONFIG_WLAN_VENDOR_ATMEL) += atmel/

