From 6d397ec8dd850b1843642cd73b3f5e0eeb610e7c Mon Sep 17 00:00:00 2001
From: Qiang Yu <yuq825@gmail.com>
Date: Tue, 19 Jun 2018 13:51:17 +0800
Subject: [PATCH 041/169] include:uapi:drm_fourcc: add ARM tiled format
 modifier

Signed-off-by: Qiang Yu <yuq825@gmail.com>

Signed-off-by: The-going <48602507+The-going@users.noreply.github.com>
---
 include/uapi/drm/drm_fourcc.h | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/include/uapi/drm/drm_fourcc.h b/include/uapi/drm/drm_fourcc.h
index fc0c1454d..0d67aad29 100644
--- a/include/uapi/drm/drm_fourcc.h
+++ b/include/uapi/drm/drm_fourcc.h
@@ -1143,6 +1143,15 @@ drm_fourcc_canonicalize_nvidia_format_mod(__u64 modifier)
  */
 #define AFRC_FORMAT_MOD_LAYOUT_SCAN (1ULL << 8)
 
+/*
+ * ARM tiled format
+ *
+ * This is used by ARM Mali Utgard/Midgard GPU. It divides buffer into
+ * 16x16 pixel blocks. Blocks are stored linearly in order, but pixels
+ * in the block are reordered.
+ */
+#define DRM_FORMAT_MOD_ARM_TILED fourcc_mod_code(ARM, 1)
+
 /*
  * Arm 16x16 Block U-Interleaved modifier
  *
-- 
2.35.3

