From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Philippe Simons <simons.philippe@gmail.com>
Date: Mon, 29 Sep 2025 20:10:53 +0200
Subject: drm:msm remove DRIVER_SYNCOBJ_TIMELINE

---
 drivers/gpu/drm/msm/msm_drv.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/gpu/drm/msm/msm_drv.c b/drivers/gpu/drm/msm/msm_drv.c
index 111111111111..222222222222 100644
--- a/drivers/gpu/drm/msm/msm_drv.c
+++ b/drivers/gpu/drm/msm/msm_drv.c
@@ -827,7 +827,6 @@ static const struct file_operations fops = {
 		DRIVER_GEM_GPUVA | \
 		DRIVER_RENDER | \
 		DRIVER_SYNCOBJ | \
-		DRIVER_SYNCOBJ_TIMELINE | \
 		0 )
 
 #define DRIVER_FEATURES_KMS ( \
-- 
Armbian

