diff --git a/common/cmd_fat.c b/common/cmd_fat.c
index f4b070b..f1c6cb1 100644
--- a/common/cmd_fat.c
+++ b/common/cmd_fat.c
@@ -300,8 +300,8 @@ int do_fat_cfgload(cmd_tbl_t *cmdtp, int flag, int argc, char *argv[])
                 "   then echo $msgload FAT;" \
                 "   else if ext4load mmc 0:1 5FFF0000 /boot/boot.ini;" \
                 "   then echo $msgload ext4 0:1;" \
-                "   else if ext4load mmc 0:2 5FFF0000 /boot/boot.ini;" \
-                "   then echo $msgload ext4 0:2;" \
+                "   else if ext4load mmc 0:1 5FFF0000 boot.ini;" \
+                "   then echo $msgload ext4 0:1;" \
                 "   fi;fi;fi", 0);
     
     if((filesize = getenv_ulong("filesize", 16, 0)))    {
