diff --git a/drivers/media/rc/keymaps/Makefile b/drivers/media/rc/keymaps/Makefile
index 793df376e15f..42836a600295 100644
--- a/drivers/media/rc/keymaps/Makefile
+++ b/drivers/media/rc/keymaps/Makefile
@@ -1,4 +1,5 @@
-obj-$(CONFIG_RC_MAP) += rc-adstech-dvb-t-pci.o \
+obj-$(CONFIG_RC_MAP) += rc-a5x.o \
+			rc-adstech-dvb-t-pci.o \
 			rc-alink-dtu-m.o \
 			rc-anysee.o \
 			rc-apac-viewcomp.o \
diff --git a/drivers/media/rc/keymaps/rc-a5x.c b/drivers/media/rc/keymaps/rc-a5x.c
new file mode 100644
index 000000000000..1871f8ec45ce
--- /dev/null
+++ b/drivers/media/rc/keymaps/rc-a5x.c
@@ -0,0 +1,53 @@
+/* Keytable for A5X IR Remote Controller
+ *
+ * Copyright (c) 2019 madmalkav
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ * NOTE: the remote have a MOUSE button, I have mapped it for MUTE
+ */
+
+#include <media/rc-map.h>
+#include <linux/module.h>
+
+static struct rc_map_table a5x[] = {
+	{ 0x8081, KEY_POWER },
+	{ 0x8083, KEY_MENU },
+	{ 0x8038, KEY_UP },
+	{ 0x8040, KEY_DOWN },
+	{ 0x8037, KEY_LEFT },
+	{ 0x8039, KEY_RIGHT },
+	{ 0x8013, KEY_OK },
+	{ 0x8089, KEY_VOLUMEDOWN },
+	{ 0x8048, KEY_MUTE },
+	{ 0x8087, KEY_VOLUMEUP },
+	{ 0x8027, KEY_BACK },
+	{ 0x8073, KEY_HOME },
+};
+
+static struct rc_map_list a5x_map = {
+	.map = {
+		.scan     = a5x,
+		.size     = ARRAY_SIZE(a5x),
+		.rc_proto = RC_PROTO_NEC,
+		.name     = RC_MAP_A5X,
+	}
+};
+
+static int __init init_rc_map_a5x(void)
+{
+	return rc_map_register(&a5x_map);
+}
+
+static void __exit exit_rc_map_a5x(void)
+{
+	rc_map_unregister(&a5x_map);
+}
+
+module_init(init_rc_map_a5x)
+module_exit(exit_rc_map_a5x)
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("madmalkav");
\ No newline at end of file
diff --git a/include/media/rc-map.h b/include/media/rc-map.h
index 6c951a1e8af0..1fdfff75dab8 100644
--- a/include/media/rc-map.h
+++ b/include/media/rc-map.h
@@ -157,6 +157,7 @@ struct rc_map *rc_map_get(const char *name);
 
 /* Names of the several keytables defined in-kernel */
 
+#define RC_MAP_A5X                       "rc-a5x"
 #define RC_MAP_ADSTECH_DVB_T_PCI         "rc-adstech-dvb-t-pci"
 #define RC_MAP_ALINK_DTU_M               "rc-alink-dtu-m"
 #define RC_MAP_ANYSEE                    "rc-anysee"
