diff --git a/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dtsi b/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dtsi
index 53f4c1f66..3c3b3e933 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dtsi
+++ b/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dtsi
@@ -720,6 +720,7 @@ &mmc1 {
 	non-removable;
 	post-power-on-delay-ms = <1>; /* wifi power is always on */
 	status = "okay";
+	keep-power-in-suspend;
 
 	rtl8723cs: wifi@1 {
 		reg = <1>;
