From eab277e9ca1f5558dbbf9820dbf6cf4edf4d4c39 Mon Sep 17 00:00:00 2001
From: The-going <48602507+The-going@users.noreply.github.com>
Date: Tue, 28 Feb 2023 13:39:51 +0300
Subject: [PATCH 153/153] mfd: sunxi-ac200: fix error initialization

---
 drivers/mfd/sunxi-ac200.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/drivers/mfd/sunxi-ac200.c b/drivers/mfd/sunxi-ac200.c
index 368a54587..b784a82f3 100644
--- a/drivers/mfd/sunxi-ac200.c
+++ b/drivers/mfd/sunxi-ac200.c
@@ -144,7 +144,7 @@ static int ac200_i2c_probe(struct i2c_client *i2c,
 	return 0;
 }
 
-static int ac200_i2c_remove(struct i2c_client *i2c)
+static void ac200_i2c_remove(struct i2c_client *i2c)
 {
 	struct ac200_dev *ac200 = i2c_get_clientdata(i2c);
 
@@ -152,8 +152,6 @@ static int ac200_i2c_remove(struct i2c_client *i2c)
 
 	mfd_remove_devices(&i2c->dev);
 	regmap_del_irq_chip(i2c->irq, ac200->regmap_irqc);
-
-	return 0;
 }
 
 static const struct i2c_device_id ac200_ids[] = {
-- 
2.35.3

