diff --git a/board/hardkernel/odroidc/cfgload.c b/board/hardkernel/odroidc/cfgload.c
index cbe1871..36c06e3 100644
--- a/board/hardkernel/odroidc/cfgload.c
+++ b/board/hardkernel/odroidc/cfgload.c
@@ -33,8 +33,8 @@ static char *load_cmd =
     " then echo Loading boot.ini from mmc0:1 (vfat);"
     " else if ext4load mmc 0:1 $loadaddr /boot/boot.ini;"
     " then echo Loading boot.ini from mmc0:1 (ext4);"
-    " else if ext4load mmc 0:2 $loadaddr /boot/boot.ini;"
-    " then echo Loading boot.init from mmc0:2 (ext4);"
+    " else if ext4load mmc 0:1 $loadaddr boot.ini;"
+    " then echo Loading boot.ini from mmc0:1 (ext4);"
     " fi;fi;fi";
 
 int do_fat_cfgload(cmd_tbl_t * cmdtp, int flag, int argc, char *argv[])
