From 35596023f707ed6f8c4df091e49c22eca6407a81 Mon Sep 17 00:00:00 2001
From: Philippe Simons <simons.philippe@gmail.com>
Date: Mon, 29 Sep 2025 20:10:53 +0200
Subject: [PATCH 24/49] drm:msm remove DRIVER_SYNCOBJ_TIMELINE

---
 drivers/gpu/drm/msm/msm_drv.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/gpu/drm/msm/msm_drv.c b/drivers/gpu/drm/msm/msm_drv.c
index e5ab1e28851d..99bb3f9a0b0c 100644
--- a/drivers/gpu/drm/msm/msm_drv.c
+++ b/drivers/gpu/drm/msm/msm_drv.c
@@ -823,7 +823,6 @@ static const struct file_operations fops = {
 		DRIVER_GEM_GPUVA | \
 		DRIVER_RENDER | \
 		DRIVER_SYNCOBJ | \
-		DRIVER_SYNCOBJ_TIMELINE | \
 		0 )
 
 #define DRIVER_FEATURES_KMS ( \
-- 
2.43.0

