diff --git a/scripts/package/builddeb b/scripts/package/builddeb
index 3c79b903..bbc054f2
--- a/scripts/package/builddeb
+++ b/scripts/package/builddeb
@@ -207,6 +207,7 @@ if [ "\$(grep nand /proc/partitions)" != "" ] && [ "\$(grep mmc /proc/partitions
 	rm -f /$installed_image_path
 else
 	ln -sf $(basename $installed_image_path) /boot/$image_name || mv /$installed_image_path /boot/$image_name
+	mkimage -A arm -O linux -T kernel -C none -a 0x00208000 -e 0x00208000 -n "Linux kernel" -d /boot/$image_name /boot/uImage
 fi
 touch /boot/.next
 exit 0
