From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Emil Renner Berthing <kernel@esmil.dk>
Date: Sun, 3 Jul 2022 21:01:11 +0200
Subject: power: reset: tps65086: Allow building as a module

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>
---
 drivers/power/reset/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/power/reset/Kconfig b/drivers/power/reset/Kconfig
index a8c46ba5878f..c3a842f185e3 100644
--- a/drivers/power/reset/Kconfig
+++ b/drivers/power/reset/Kconfig
@@ -205,7 +205,7 @@ config POWER_RESET_ST
 	  Reset support for STMicroelectronics boards.
 
 config POWER_RESET_TPS65086
-	bool "TPS65086 restart driver"
+	tristate "TPS65086 restart driver"
 	depends on MFD_TPS65086
 	help
 	  This driver adds support for resetting the TPS65086 PMIC on restart.
-- 
Armbian

