diff --git a/scripts/package/builddeb b/scripts/package/builddeb
index 750d70e..8fc16c9 100755
--- a/scripts/package/builddeb
+++ b/scripts/package/builddeb
@@ -258,7 +258,8 @@ mountpoint -q /boot || mount /boot
 cp /tmp/uImage /boot/uImage
 rm -f /$installed_image_path
 else
-ln -sf $(basename $installed_image_path) /boot/Image > /dev/null 2>&1 || mv /$installed_image_path /boot/Image
+rm -f /boot/zImage
+mkimage -A arm64 -O linux -T kernel -C none -a 0x1080000 -e 0x1080000 -n "Linux kernel" -d /boot/$(basename $installed_image_path) /boot/Image
 fi
 touch /boot/.next
 exit 0
