From 8fa74f89cd0cbea242c9aae234a99400a6357038 Mon Sep 17 00:00:00 2001
From: Stefan Mavrodiev <stefan@olimex.com>
Date: Fri, 17 May 2019 13:17:50 +0300
Subject: [PATCH 1/2] Fix printf error

Signed-off-by: Stefan Mavrodiev <stefan@olimex.com>
---
 src/armsoc_exa.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/armsoc_exa.c b/src/armsoc_exa.c
index c58db80..494d43d 100644
--- a/src/armsoc_exa.c
+++ b/src/armsoc_exa.c
@@ -260,7 +260,7 @@ ModifyUnAccelPixmapHeader(struct ARMSOCPixmapPrivRec *priv, PixmapPtr pPixmap, i
 		priv->unaccel = malloc(datasize);
 
 		if (!priv->unaccel) {
-			ERROR_MSG("failed to allocate %d bytes mem",
+			ERROR_MSG("failed to allocate %ld bytes mem",
 					datasize);
 			priv->unaccel_size = 0;
 			return FALSE;
-- 
2.17.1

