From 5f2a1b8dd2636d7712d8d01fc0c77fe5cce6fe97 Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megous@megous.com>
Date: Mon, 30 Nov 2020 06:13:03 +0100
Subject: [PATCH 180/446] usb: typec: anx7688: Loudly report connection failure

Kinda useful.

Signed-off-by: Ondrej Jirman <megous@megous.com>
---
 drivers/usb/typec/anx7688.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/usb/typec/anx7688.c b/drivers/usb/typec/anx7688.c
index 0db7d93c4..63dd6ab42 100644
--- a/drivers/usb/typec/anx7688.c
+++ b/drivers/usb/typec/anx7688.c
@@ -537,6 +537,7 @@ static int anx7688_connect(struct anx7688 *anx7688)
 	anx7688->vconn_on = false;
 err_poweroff:
         anx7688_power_disable(anx7688);
+        dev_err(anx7688->dev, "OCM configuration failed\n");
         return ret;
 #endif
 }
-- 
2.31.1

