From 908176615a6aa710563883715da75e50e7305179 Mon Sep 17 00:00:00 2001
From: Paolo Sabatino <paolo.sabatino@gmail.com>
Date: Sat, 4 May 2024 15:07:27 +0200
Subject: [PATCH] add reset properties for usb otg/ehci

---
 arch/arm/dts/rk322x.dtsi | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/arch/arm/dts/rk322x.dtsi b/arch/arm/dts/rk322x.dtsi
index 8eed9e3a92..ffe503e5db 100644
--- a/arch/arm/dts/rk322x.dtsi
+++ b/arch/arm/dts/rk322x.dtsi
@@ -799,6 +799,8 @@
 		g-tx-fifo-size = <256 128 128 64 32 16>;
 		phys = <&u2phy0_otg>;
 		phy-names = "usb2-phy";
+		resets = <&cru SRST_USBOTG>;
+		reset-names = "dwc2";
 		status = "disabled";
 	};
 
@@ -809,6 +811,8 @@
 		clocks = <&cru HCLK_HOST0>, <&u2phy0>;
 		phys = <&u2phy0_host>;
 		phy-names = "usb";
+		resets = <&cru SRST_USBHOST0>;
+		reset-names = "ehci";
 		status = "disabled";
 	};
 
@@ -829,6 +833,8 @@
 		clocks = <&cru HCLK_HOST1>, <&u2phy1>;
 		phys = <&u2phy1_otg>;
 		phy-names = "usb";
+		resets = <&cru SRST_USBHOST1>;
+		reset-names = "ehci";
 		status = "disabled";
 	};
 
@@ -849,6 +855,8 @@
 		clocks = <&cru HCLK_HOST2>, <&u2phy1>;
 		phys = <&u2phy1_host>;
 		phy-names = "usb";
+		resets = <&cru SRST_USBHOST2>;
+		reset-names = "ehci";
 		status = "disabled";
 	};
 
-- 
2.34.1

