From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: The-going <48602507+The-going@users.noreply.github.com>
Date: Wed, 2 Feb 2022 11:56:51 +0300
Subject: drv:spi:spidev Add armbian spi-dev compatible

---
 drivers/spi/spidev.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/spi/spidev.c b/drivers/spi/spidev.c
index 4e0703304f1d..9c8256a764d6 100644
--- a/drivers/spi/spidev.c
+++ b/drivers/spi/spidev.c
@@ -701,10 +701,11 @@ static const struct class spidev_class = {
  * The spi device ids are expected to match the device names of the
  * spidev_dt_ids array below. Both arrays are kept in the same ordering.
  */
 static const struct spi_device_id spidev_spi_ids[] = {
 	{ .name = /* abb */ "spi-sensor" },
+	{ .name = /* armbian */ "spi-dev" },
 	{ .name = /* cisco */ "spi-petra" },
 	{ .name = /* dh */ "dhcom-board" },
 	{ .name = /* elgin */ "jg10309-01" },
 	{ .name = /* gocontroll */ "moduline-module-slot"},
 	{ .name = /* lineartechnology */ "ltc2488" },
@@ -729,15 +730,17 @@ static int spidev_of_check(struct device *dev)
 {
 	if (device_property_match_string(dev, "compatible", "spidev") < 0)
 		return 0;
 
 	dev_err(dev, "spidev listed directly in DT is not supported\n");
+
 	return -EINVAL;
 }
 
 static const struct of_device_id spidev_dt_ids[] = {
 	{ .compatible = "abb,spi-sensor", .data = &spidev_of_check },
+	{ .compatible = "armbian,spi-dev", .data = &spidev_of_check },
 	{ .compatible = "cisco,spi-petra", .data = &spidev_of_check },
 	{ .compatible = "dh,dhcom-board", .data = &spidev_of_check },
 	{ .compatible = "elgin,jg10309-01", .data = &spidev_of_check },
 	{ .compatible = "gocontroll,moduline-module-slot", .data = &spidev_of_check},
 	{ .compatible = "lineartechnology,ltc2488", .data = &spidev_of_check },
@@ -756,10 +759,11 @@ MODULE_DEVICE_TABLE(of, spidev_dt_ids);
 
 /* Dummy SPI devices not to be used in production systems */
 static int spidev_acpi_check(struct device *dev)
 {
 	dev_warn(dev, "do not use this driver in production systems!\n");
+	dev_info(dev, "Use a compatible alias string like spi-dev in DT\n");
 	return 0;
 }
 
 static const struct acpi_device_id spidev_acpi_ids[] = {
 	/*
-- 
Armbian

