diff --git a/arch/arm/boot/dts/armada-388-helios4.dts b/arch/arm/boot/dts/armada-388-helios4.dts
index 295f962..9c536d7 100644
--- a/arch/arm/boot/dts/armada-388-helios4.dts
+++ b/arch/arm/boot/dts/armada-388-helios4.dts
@@ -276,11 +276,6 @@
 				status = "okay";
 				vmmc = <&reg_3p3v>;
 				wp-inverted;
-				max-frequency = <50000000>;
-				cap-sd-highspeed;
-				sd-uhs-sdr12;
-				sd-uhs-sdr25;
-				sd-uhs-sdr50;
 			};
 
 			usb@58000 {
