diff --git a/arch/arm/boot/dts/armada-388-helios4.dts b/arch/arm/boot/dts/armada-388-helios4.dts
index ec7a5ce..1075535 100644
--- a/arch/arm/boot/dts/armada-388-helios4.dts
+++ b/arch/arm/boot/dts/armada-388-helios4.dts
@@ -131,13 +131,17 @@
 	};
 
 	fan1: j10-pwm {
-		compatible = "pwm-fan";
-		pwms = <&gpio1 9 3800>;
+		compatible = "gpio-fan";
+		gpios = <&gpio1 9 GPIO_ACTIVE_HIGH>;
+		gpio-fan,speed-map = <	 0 0
+				      3800 1>;
 	};
 
 	fan2: j17-pwm {
-		compatible = "pwm-fan";
-		pwms = <&gpio1 23 3800>;
+		compatible = "gpio-fan";
+		gpios = <&gpio1 23 GPIO_ACTIVE_HIGH>;
+		gpio-fan,speed-map = <	 0 0
+				      3800 1>;
 	};
 
 	usb2_phy: usb2-phy {
