diff --git a/include/configs/udoo_neo.h b/include/configs/udoo_neo.h
index 4a4eb51..00c011c 100644
--- a/include/configs/udoo_neo.h
+++ b/include/configs/udoo_neo.h
@@ -187,8 +187,13 @@
 	UPDATE_M4_ENV \
 	M4_STARTUP_ENV \
 	CONFIG_VIDEO_MODE \
-	"script=uEnv.txt\0" \
-	"image=/zImage\0" \
+	"script=boot.scr\0" \
+	"board=Neo\0" \
+	"boot_prefixes=/ /boot/\0" \
+	"boot_a_script=ext2load mmc ${mmcdev}:${mmcpart} ${loadaddr} ${prefix}${script} || fatload mmc ${mmcdev}:${mmcpart} ${loadaddr} ${prefix}${script} \0" \
+	"scan_dev_for_boot=echo Scanning mmc ; for prefix in ${boot_prefixes}; do run scan_dev_for_scripts; done\0" \
+	"scan_dev_for_scripts=if test -e ${mmcdev}:${mmcpart} ${prefix}${script}; then echo Found U-Boot script ${prefix}${script}; run boot_a_script; else echo Checking here ${prefix} ... not found; fi;\0" \
+	"image=zImage\0" \
 	"console=ttymxc0\0" \
 	"fdt_high=0xffffffff\0" \
 	"initrd_high=0xffffffff\0" \
@@ -204,9 +208,9 @@
 	"mmcargs=setenv bootargs console=${console},${baudrate} " \
 		"root=${mmcroot} rootfstype=${mmcrootfstype} ${m4mmcargs} consoleblank=0\0" \
 	"loadbootscript=" \
-		"fatload mmc ${mmcdev}:${mmcpart} ${loadaddr} ${script};\0" \
+		"run scan_dev_for_boot;\0" \
 	"bootscript=echo Running bootscript from mmc ...; " \
-		"env import -t ${loadaddr} ${filesize};\0" \
+		"source\0" \
 	"loadimage=fatload mmc ${mmcdev}:${mmcpart} ${loadaddr} ${image}\0" \
 	"loadfdt=fatload mmc ${mmcdev}:${mmcpart} ${fdt_addr} ${fdt_file}\0" \
 	"mmcboot=echo Booting from mmc ...; " \
