diff --git a/common_drivers/drivers/drm/meson_fbdev.c b/common_drivers/drivers/drm/meson_fbdev.c
index 3d1cb62aab44..3718a39888f2 100644
--- a/common_drivers/drivers/drm/meson_fbdev.c
+++ b/common_drivers/drivers/drm/meson_fbdev.c
@@ -447,7 +447,6 @@ int am_meson_drm_fb_helper_set_par(struct fb_info *info)
 			MESON_DRM_FBDEV("%s reallocate success.\n", __func__);
 		}
 	}
-	drm_wait_one_vblank(fb_helper->dev, 0);
 	DRM_INFO("fb_set_par: %s OUT\n", __func__);
 
 	return 0;
