diff --git a/drivers/base/devres.c b/drivers/base/devres.c
index e038e2b..37a2182 100644
--- a/drivers/base/devres.c
+++ b/drivers/base/devres.c
@@ -33,7 +33,8 @@ struct devres {
 	 * Thus we use ARCH_KMALLOC_MINALIGN here and get exactly the same
 	 * buffer alignment as if it was allocated by plain kmalloc().
 	 */
-	u8 __aligned(ARCH_KMALLOC_MINALIGN) data[];
+	//u8 __aligned(ARCH_KMALLOC_MINALIGN) data[];
+	unsigned long long              data[];
 };
 
 struct devres_group {
