diff --git a/include/configs/s500.h b/include/configs/s500.h
index 3070548..a16fe2f 100755
--- a/include/configs/s500.h
+++ b/include/configs/s500.h
@@ -32,7 +32,9 @@
 	"loadkernel=fatload ${devtype} ${devpart} ${kernel_addr_r} uImage\0"	\
 	"loadramdisk=fatload ${devtype} ${devpart} ${ramdisk_addr_r} ramdisk.img\0" \
 	"loadfdt=fatload ${devtype} ${devpart} ${fdt_addr_r} kernel.dtb\0" \
-	"loadbootscr=fatload ${devtype} ${devpart} ${scriptaddr} ${bootscr} \0" \
+	"loadbootscr=ext4load ${devtype} ${devpart} ${scriptaddr} /boot/${bootscr} || " \
+	"fatload ${devtype} ${devpart} ${scriptaddr} ${bootscr} || " \
+	"ext4load ${devtype} ${devpart} ${scriptaddr} ${bootscr}\0" \
 	"loadbootenv=fatload ${devtype} ${devpart} ${scriptaddr} ${bootenv} \0" \
 	"mmcargs=setenv devtype mmc; setenv devpart ${mmcdev}:${bootpart}\0" \
 	"emmcargs=setenv devtype mmc; setenv devpart ${emmcdev}:${bootpart}\0"	\
