From 56e4952840fb0668d9f4a15a85c5eb4c2d593abd Mon Sep 17 00:00:00 2001
From: The-going <48602507+The-going@users.noreply.github.com>
Date: Wed, 15 Jan 2025 11:52:31 +0300
Subject: [PATCH] arm64: dts: orangepi-zero2w: add uart 2,3,4,5

---
 .../allwinner/sun50i-h618-orangepi-zero2w.dts | 51 +++++++++++++++++++
 1 file changed, 51 insertions(+)

diff --git a/arch/arm64/boot/dts/allwinner/sun50i-h618-orangepi-zero2w.dts b/arch/arm64/boot/dts/allwinner/sun50i-h618-orangepi-zero2w.dts
index 777401341970..94361afaa6c3 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-h618-orangepi-zero2w.dts
+++ b/arch/arm64/boot/dts/allwinner/sun50i-h618-orangepi-zero2w.dts
@@ -19,6 +19,10 @@ / {
 
 	aliases {
 		serial0 = &uart0;
+		serial2 = &uart2;
+		serial3 = &uart3;
+		serial4 = &uart4;
+		serial5 = &uart5;
 	};
 
 	chosen {
@@ -153,6 +157,30 @@ &pio {
 	vcc-ph-supply = <&reg_dldo1>;	/* internally via VCC-IO */
 	vcc-pi-supply = <&reg_dldo1>;
 
+	/omit-if-no-ref/
+	uart2_pi_pins: uart2-pi-pins {
+		pins = "PI5", "PI6";
+		function = "uart2";
+	};
+
+	/omit-if-no-ref/
+	uart3_pi_pins: uart3-pi-pins {
+		pins = "PI9", "PI10";
+		function = "uart3";
+	};
+
+	/omit-if-no-ref/
+	uart4_pi_pins: uart4-pi-pins {
+		pins = "PI13", "PI14";
+		function = "uart4";
+	};
+
+	/omit-if-no-ref/
+	uart5_ph_pins: uart5-ph-pins {
+		pins = "PH2", "PH3";
+		function = "uart5";
+	};
+
 	/omit-if-no-ref/
 	pwm1_ph_pin: pwm1-ph-pin {
 		pins = "PH3";
@@ -260,6 +288,29 @@ &uart0 {
 	status = "okay";
 };
 
+&uart2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart2_pi_pins>;
+	status = "disabled";
+};
+
+&uart3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart3_pi_pins>;
+	status = "disabled";
+};
+
+&uart4 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart4_pi_pins>;
+	status = "disabled";
+};
+
+&uart5 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart5_ph_pins>;
+	status = "disabled";
+
 &usbotg {
 	/*
 	 * PHY0 pins are connected to a USB-C socket, but a role switch
-- 
2.35.3

