diff --git a/arch/arm/boot/dts/armada-388-helios4.dts b/arch/arm/boot/dts/armada-388-helios4.dts
index 705adfa8c..d5afbfc53 100644
--- a/arch/arm/boot/dts/armada-388-helios4.dts
+++ b/arch/arm/boot/dts/armada-388-helios4.dts
@@ -84,6 +84,18 @@
 		};
 	};
 
+	gpio-keys {
+		compatible = "gpio-keys";
+		pinctrl-0 = <&microsom_phy0_int_pins>;
+
+		wol {
+			label = "Wake-On-LAN";
+			linux,code = <KEY_WAKEUP>;
+			gpios = <&gpio0 18 GPIO_ACTIVE_LOW>;
+			wakeup-source;
+		};
+	};
+
 	io-leds {
 		compatible = "gpio-leds";
 		sata1-led {
