diff --git a/arch/arm/mach-rockchip/boot_rkimg.c b/arch/arm/mach-rockchip/boot_rkimg.c
index cb5db30dba9..e26a9bdcb79 100644
--- a/arch/arm/mach-rockchip/boot_rkimg.c
+++ b/arch/arm/mach-rockchip/boot_rkimg.c
@@ -97,13 +97,15 @@ static void boot_devtype_init(void)
 	char *devtype = NULL, *devnum = NULL;
 	char *src = "scan";
 	static int done;	/* static */
 	int ret;
+	bool need_reinit;
 
-	if (done)
+	need_reinit = !done || !env_get("devtype") || !env_get("devnum");
+	if (!need_reinit)
 		return;
 
 #ifdef CONFIG_MP_BOOT
 	mpb_post(0);
 #endif
 
 	/* configuration */
