From 8ad90bf0ea85aa771f31abb9ab2279bbaffbceac Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megi@xff.cz>
Date: Sun, 30 Jul 2023 10:30:06 +0200
Subject: [PATCH 414/464] bes2600: Don't kill the association 3s after joining

Signed-off-by: Ondrej Jirman <megi@xff.cz>
---
 drivers/staging/bes2600/sta.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/staging/bes2600/sta.c b/drivers/staging/bes2600/sta.c
index 4fa7e4ad9999..9e96085fdd6d 100644
--- a/drivers/staging/bes2600/sta.c
+++ b/drivers/staging/bes2600/sta.c
@@ -2361,6 +2361,8 @@ void bes2600_join_work(struct work_struct *work)
 			priv->join_status = BES2600_JOIN_STATUS_STA;
 			atomic_set(&priv->connect_in_process, 1);
 
+			cancel_delayed_work_sync(&priv->join_timeout);
+
 			/* Due to beacon filtering it is possible that the
 			 * AP's beacon is not known for the mac80211 stack.
 			 * Disable filtering temporary to make sure the stack
-- 
2.34.1

