diff --git a/arch/arm/Kconfig b/arch/arm/Kconfig
index 84462f3..3291f9f
--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -1112,6 +1112,8 @@ config ARCH_ROCKCHIP
 	select DM_USB if USB
 	select DM_PWM
 	select DM_REGULATOR
+	select OF_LIBFDT
+	select OF_LIBFDT_OVERLAY
 	select CMD_ROCKUSB if USB_GADGET_DOWNLOAD
 	select ENABLE_ARM_SOC_BOOT0_HOOK
 	imply CMD_FASTBOOT
diff --git a/include/configs/rk3328_common.h b/include/configs/rk3328_common.h
index 6be9ad7..f553425 100644
--- a/include/configs/rk3328_common.h
+++ b/include/configs/rk3328_common.h
@@ -57,6 +57,7 @@
 	"fdt_addr_r=0x08300000\0" \
 	"kernel_addr_r=0x00280000\0" \
 	"ramdisk_addr_r=0x0a200000\0"
+#define FDTFILE "rockchip/" CONFIG_DEFAULT_DEVICE_TREE ".dtb"
 
 #include <config_distro_bootcmd.h>
 #define CONFIG_EXTRA_ENV_SETTINGS \
@@ -64,6 +65,7 @@
 	RKIMG_DET_BOOTDEV \
 	"partitions=" PARTS_DEFAULT \
 	ROCKCHIP_DEVICE_SETTINGS \
+	"fdtfile=" FDTFILE "\0" \
 	BOOTENV
 
 #endif
