From daed33d0944439dd9239057763b5e03bd9dd379f Mon Sep 17 00:00:00 2001
From: schwar3kat <61094841+schwar3kat@users.noreply.github.com>
Date: Sun, 3 Apr 2022 00:20:50 +1300
Subject: [PATCH] fix-gpio-kconfig remove if EXPERT to allow normal build

Signed-off-by: schwar3kat <61094841+schwar3kat@users.noreply.github.com>
---
 drivers/gpio/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/gpio/Kconfig b/drivers/gpio/Kconfig
index 947474f6a..0458a96a7 100644
--- a/drivers/gpio/Kconfig
+++ b/drivers/gpio/Kconfig
@@ -59,7 +59,7 @@ config DEBUG_GPIO
 	  that are most common when setting up new platforms or boards.
 
 config GPIO_SYSFS
-	bool "/sys/class/gpio/... (sysfs interface)" if EXPERT
+	bool "/sys/class/gpio/... (sysfs interface)"
 	depends on SYSFS
 	select GPIO_CDEV # We need to encourage the new ABI
 	help
-- 


