From 2dda569d0b7193631cc21932b2ca5421b6787d88 Mon Sep 17 00:00:00 2001
From: Jernej Skrabec <jernej.skrabec@gmail.com>
Date: Fri, 16 Jan 2026 09:05:06 -0500
Subject: [PATCH] HACK: sunxi: h616 gpu enable

Signed-off-by: Jernej Skrabec <jernej.skrabec@gmail.com>
---
 arch/arm/mach-sunxi/clock_sun50i_h6.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/arch/arm/mach-sunxi/clock_sun50i_h6.c b/arch/arm/mach-sunxi/clock_sun50i_h6.c
index 80004f13a1e..d2986ec0b29 100644
--- a/arch/arm/mach-sunxi/clock_sun50i_h6.c
+++ b/arch/arm/mach-sunxi/clock_sun50i_h6.c
@@ -40,6 +40,8 @@ void clock_init_safe(void)
 		setbits_le32(prcm + CCU_PRCM_PLL_LDO_CFG, 0xA7000000);
 		/* set PLL VDD LDO output to 1.14 V */
 		setbits_le32(prcm + CCU_PRCM_PLL_LDO_CFG, 0x60000);
+		/* enable GPU */
+		writel(0, 0x7010254);
 	}
 
 	clock_set_pll1(408000000);
-- 
2.51.0

