diff --git a/drivers/mmc/host/sunxi-mmc-debug.c b/drivers/mmc/host/sunxi-mmc-debug.c
index 7d43cd1d..2724c7e7 100755
--- a/drivers/mmc/host/sunxi-mmc-debug.c
+++ b/drivers/mmc/host/sunxi-mmc-debug.c
@@ -325,7 +325,7 @@ static ssize_t dump_clk_dly_show(struct device *dev, struct device_attribute *at
 	return p-buf;
 }
 
-
+/*
 
 int mmc_create_sys_fs(struct sunxi_mmc_host* host,struct platform_device *pdev)
 {
@@ -388,5 +388,5 @@ void mmc_remove_sys_fs(struct sunxi_mmc_host* host,struct platform_device *pdev)
 	device_remove_file(&pdev->dev, &host->dump_clk_dly);
 }
 
-
+*/
 
diff --git a/drivers/mmc/host/sunxi-mmc-debug.h b/drivers/mmc/host/sunxi-mmc-debug.h
index c855c8ae..43c60e5f 100755
--- a/drivers/mmc/host/sunxi-mmc-debug.h
+++ b/drivers/mmc/host/sunxi-mmc-debug.h
@@ -19,8 +19,8 @@ void sunxi_mmc_dumphex32(struct sunxi_mmc_host* host, char* name, char* base, in
 void sunxi_mmc_dump_des(struct sunxi_mmc_host* host, char* base, int len);
 
 
-int mmc_create_sys_fs(struct sunxi_mmc_host* host,struct platform_device *pdev);
-void mmc_remove_sys_fs(struct sunxi_mmc_host* host,struct platform_device *pdev);
+/* int mmc_create_sys_fs(struct sunxi_mmc_host* host,struct platform_device *pdev);
+void mmc_remove_sys_fs(struct sunxi_mmc_host* host,struct platform_device *pdev); */
 void dump_reg(struct sunxi_mmc_host *host);
 
 #endif
diff --git a/drivers/mmc/host/sunxi-mmc.c b/drivers/mmc/host/sunxi-mmc.c
index b5a69639..4097a4e0 100755
--- a/drivers/mmc/host/sunxi-mmc.c
+++ b/drivers/mmc/host/sunxi-mmc.c
@@ -1522,11 +1522,11 @@ static int sunxi_mmc_probe(struct platform_device *pdev)
 	if (ret)
 		goto error_free_dma;
 
-	ret = mmc_create_sys_fs(host,pdev);
+/*	ret = mmc_create_sys_fs(host,pdev);
 	if(ret){
 		dev_err(&pdev->dev, "create sys fs failed\n");
 		goto error_free_dma;
-	}
+	} */
 
 	dev_info(&pdev->dev, "base:0x%p irq:%u\n", host->reg_base, host->irq);
 	platform_set_drvdata(pdev, mmc);
@@ -1548,7 +1548,7 @@ static int sunxi_mmc_remove(struct platform_device *pdev)
 	disable_irq(host->irq);
 	sunxi_mmc_reset_host(host);
 
-	mmc_remove_sys_fs(host,pdev);
+/*	mmc_remove_sys_fs(host,pdev); */
 
 	if (!IS_ERR(mmc->supply.vdmmc))
 			regulator_disable(mmc->supply.vdmmc);
diff --git a/drivers/mmc/host/sunxi-smhc.c b/drivers/mmc/host/sunxi-smhc.c
index 6474794b..9422c012 100644
--- a/drivers/mmc/host/sunxi-smhc.c
+++ b/drivers/mmc/host/sunxi-smhc.c
@@ -1537,11 +1537,11 @@ static int sunxi_mmc_probe(struct platform_device *pdev)
 	if (ret)
 		goto error_free_dma;
 
-	ret = mmc_create_sys_fs(host,pdev);
+/*	ret = mmc_create_sys_fs(host,pdev);
 	if(ret){
 		dev_err(&pdev->dev, "create sys fs failed\n");
 		goto error_free_dma;
-	}
+	} */
 	
 	dev_info(&pdev->dev, "base:0x%p irq:%u\n", host->reg_base, host->irq);
 	platform_set_drvdata(pdev, mmc);
@@ -1563,7 +1563,7 @@ static int sunxi_mmc_remove(struct platform_device *pdev)
 	disable_irq(host->irq);
 //	sunxi_mmc_reset_host(host);
 
-	mmc_remove_sys_fs(host,pdev);
+/*	mmc_remove_sys_fs(host,pdev); */
 
 	if (!IS_ERR(mmc->supply.vdmmc))
 			regulator_disable(mmc->supply.vdmmc);
