diff -uNr a/arch/arm64/boot/dts/allwinner/overlay/Makefile b/arch/arm64/boot/dts/allwinner/overlay/Makefile
--- a/arch/arm64/boot/dts/allwinner/overlay/Makefile	2021-08-05 22:56:13.983779398 +0800
+++ b/arch/arm64/boot/dts/allwinner/overlay/Makefile	2021-08-05 23:21:13.534600814 +0800
@@ -38,6 +38,7 @@
 	sun50i-h6-i2c0.dtbo \
 	sun50i-h6-i2c1.dtbo \
 	sun50i-h6-i2c2.dtbo \
+	sun50i-h6-pps-gpio.dtbo \
 	sun50i-h6-ruart.dtbo \
 	sun50i-h6-spi-add-cs1.dtbo \
 	sun50i-h6-spi-jedec-nor.dtbo \
diff -uNr a/arch/arm64/boot/dts/allwinner/overlay/sun50i-h6-pps-gpio.dts b/arch/arm64/boot/dts/allwinner/overlay/sun50i-h6-pps-gpio.dts
--- a/arch/arm64/boot/dts/allwinner/overlay/sun50i-h6-pps-gpio.dts	1970-01-01 08:00:00.000000000 +0800
+++ b/arch/arm64/boot/dts/allwinner/overlay/sun50i-h6-pps-gpio.dts	2021-08-05 23:22:11.207147333 +0800
@@ -0,0 +1,29 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	compatible = "allwinner,sun50i-h6";
+
+	fragment@0 {
+		target = <&pio>;
+		__overlay__ {
+			pps_pins: pps_pins {
+				pins = "PL10";
+				function = "gpio_in";
+			};
+		};
+	};
+
+	fragment@1 {
+		target-path = "/";
+		__overlay__ {
+			pps@0 {
+				compatible = "pps-gpio";
+				pinctrl-names = "default";
+				pinctrl-0 = <&pps_pins>;
+				gpios = <&pio 11 10 0>; /* PL10 */
+				status = "okay";
+			};
+		};
+	};
+};

