diff --git a/arch/arm/dts/rk3568-radxa-rock-3-compute-module-plus.dtsi b/arch/arm/dts/rk3568-radxa-rock-3-compute-module-plus.dtsi
index e2cdda8a..15d7b41f 100644
--- a/arch/arm/dts/rk3568-radxa-rock-3-compute-module-plus.dtsi
+++ b/arch/arm/dts/rk3568-radxa-rock-3-compute-module-plus.dtsi
@@ -372,6 +372,7 @@
 				regulator-boot-on;
 				regulator-name = "vcc_3v3";
 				regulator-state-mem {
+					u-boot,dm-pre-reloc;
 					regulator-off-in-suspend;
 				};
 			};
@@ -418,4 +419,4 @@
 				<0 RK_PA2 RK_FUNC_2 &pcfg_pull_none>;
 		};
 	};
-};
\ No newline at end of file
+};
