diff --git a/board/sunxi/board.c b/board/sunxi/board.c
index 827e545032..7d343a93b2 100644
--- a/board/sunxi/board.c
+++ b/board/sunxi/board.c
@@ -581,6 +581,11 @@ void sunxi_board_init(void)
 		status_led_init();
 #endif
 
+#ifdef CONFIG_MACH_SUNXI_H3_H5
+	/* turn on power LED (PL10) on H3 and H5 boards */
+	gpio_direction_output(SUNXI_GPL(10), 1);
+#endif
+
 #ifdef CONFIG_SY8106A_POWER
 	power_failed = sy8106a_set_vout1(CONFIG_SY8106A_VOUT1_VOLT);
 #endif
