diff --git a/arch/arm64/boot/dts/allwinner/sun50i-h5-orangepi-zero-plus2.dts b/arch/arm64/boot/dts/allwinner/sun50i-h5-orangepi-zero-plus2.dts
index b6b7a56..62e5ebf 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-h5-orangepi-zero-plus2.dts
+++ b/arch/arm64/boot/dts/allwinner/sun50i-h5-orangepi-zero-plus2.dts
@@ -45,6 +45,7 @@
 #include "sun50i-h5.dtsi"
 
 #include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/sun4i-a10.h>
 
 / {
 	model = "OrangePi Zero Plus2";
@@ -64,6 +65,14 @@
 		regulator-min-microvolt = <3300000>;
 		regulator-max-microvolt = <3300000>;
 	};
+
+        wifi_pwrseq: wifi_pwrseq {
+                compatible = "mmc-pwrseq-simple";
+                pinctrl-names = "default";
+                pinctrl-0 = <&wifi_en_pin>;
+                reset-gpios = <&pio 0 9 GPIO_ACTIVE_LOW>; /* PA9 */
+                post-power-on-delay-ms = <50>;
+        };
 };
 
 &mmc0 {
@@ -75,6 +84,25 @@
 	status = "okay";
 };
 
+&mmc1 {
+        pinctrl-names = "default";
+        pinctrl-0 = <&mmc1_pins_a>;
+        vmmc-supply = <&reg_vcc3v3>;
+        vqmmc-supply = <&reg_vcc3v3>;
+        mmc-pwrseq = <&wifi_pwrseq>;
+        bus-width = <4>;
+        non-removable;
+        status = "okay";
+
+        brcmf: brcmf@1 {
+                reg = <1>;
+                compatible = "brcm,bcm4329-fmac";   
+                interrupt-parent = <&r_pio>;
+                interrupts = <0 7 IRQ_TYPE_LEVEL_LOW>; /* PL7 / EINT7 */
+                interrupt-names = "host-wake";
+        };
+};
+
 &mmc2 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&mmc2_8bit_pins>;
@@ -85,6 +113,23 @@
 	status = "okay";
 };
 
+&pio {
+        wifi_en_pin: wifi_en_pin@0 {
+                allwinner,pins = "PA9";
+                allwinner,function = "gpio_out";
+                allwinner,drive = <SUN4I_PINCTRL_10_MA>;
+                allwinner,pull = <SUN4I_PINCTRL_NO_PULL>;
+        };
+};
+
+&r_pio {
+        wifi_wake: wifi_wake@0 {
+                 allwinner,pins = "PL7";
+                allwinner,function = "irq";
+                allwinner,pull = <SUN4I_PINCTRL_PULL_UP>;
+        };
+};
+
 &uart0 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&uart0_pins_a>;
