diff --git a/arch/arm/boot/dts/sun7i-a20-olinuxino-micro.dts b/arch/arm/boot/dts/sun7i-a20-olinuxino-micro.dts
index 0b7403e..ed88c76 100644
--- a/arch/arm/boot/dts/sun7i-a20-olinuxino-micro.dts
+++ b/arch/arm/boot/dts/sun7i-a20-olinuxino-micro.dts
@@ -100,16 +100,15 @@
 	status = "okay";
 };
 
-&gmac {
+&emac {
 	pinctrl-names = "default";
-	pinctrl-0 = <&gmac_pins_mii_a>;
+	pinctrl-0 = <&emac_pins_a>,<&emac_txerr_pin>;
 	phy = <&phy1>;
-	phy-mode = "mii";
 	status = "okay";
+};
 
-	phy1: ethernet-phy@1 {
-		reg = <1>;
-	};
+&emac_sram {
+	status = "okay";
 };
 
 &i2c0 {
@@ -196,6 +195,14 @@
 	};
 };
 
+&mdio {
+	status = "okay";
+
+	phy1: ethernet-phy@1 {
+		reg = <1>;
+	};
+};
+
 &mmc0 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&mmc0_pins_a>;
@@ -229,6 +236,11 @@
 };
 
 &pio {
+	emac_txerr_pin: emac_txerr_pin@0 {
+		pins = "PA17";
+		function = "emac";
+	};
+
 	mmc3_cd_pin_olinuxinom: mmc3_cd_pin@0 {
 		pins = "PH11";
 		function = "gpio_in";
