diff --git a/drivers/net/wireless/rtl8189fs/include/autoconf.h b/drivers/net/wireless/rtl8189fs/include/autoconf.h
index 44efe42..6f40903 100644
--- a/drivers/net/wireless/rtl8189fs/include/autoconf.h
+++ b/drivers/net/wireless/rtl8189fs/include/autoconf.h
@@ -224,7 +224,7 @@
 /*
  * Debug Related Config
  */
-#define CONFIG_DEBUG /* DBG_871X, etc... */
+//#define CONFIG_DEBUG /* DBG_871X, etc... */
 
 #ifdef CONFIG_DEBUG
 #define DBG	1	// for ODM & BTCOEX debug
