diff --git a/arch/arm64/boot/dts/rockchip/rk3328-rock64.dts b/arch/arm64/boot/dts/rockchip/rk3328-rock64.dts
index 113b82cd..d0612eb1 100644
--- a/arch/arm64/boot/dts/rockchip/rk3328-rock64.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3328-rock64.dts
@@ -47,6 +47,10 @@
 	model = "PINE64 ROCK64";
 	compatible = "pine64,rock64", "rockchip,rk3328";

+	aliases {
+		ethernet0 = &gmac2io;
+	};
+
 	chosen {
 		bootargs = "rockchip_jtag earlyprintk=uart8250-32bit,0xff130000";
 	};
