diff --git a/drivers/mmc/host/dw_mmc-rockchip.c b/drivers/mmc/host/dw_mmc-rockchip.c
index 3fc65f4..274ddba 100644
--- a/drivers/mmc/host/dw_mmc-rockchip.c
+++ b/drivers/mmc/host/dw_mmc-rockchip.c
@@ -291,7 +291,7 @@ static int dw_mci_rockchip_probe(struct platform_device *pdev)
 static void dw_mci_rockchip_platfm_shutdown(struct platform_device *pdev)
 {
 	struct dw_mci *host = platform_get_drvdata(pdev);
-	struct mmc_host *mmc = host->cur_slot->mmc;
+	struct mmc_host *mmc = host->slot[0]->mmc;
 	int ret;
 
 	if(of_machine_is_compatible("rockchip,rk3288-miniarm")){
diff --git a/arch/arm/boot/dts/rk3288-miniarm.dts b/arch/arm/boot/dts/rk3288-miniarm.dts
index 5371310..e9c9eac 100644
--- a/arch/arm/boot/dts/rk3288-miniarm.dts
+++ b/arch/arm/boot/dts/rk3288-miniarm.dts
@@ -45,6 +45,9 @@
 
 / {
 	compatible = "rockchip,rk3288-miniarm", "rockchip,rk3288";
+        model = "                                ";
+        copyright = "ASUS";
+	system-id = " "; /*For ASUS PPID*/
 
 	memory {
 		device_type = "memory";
@@ -173,6 +176,27 @@
 		vin-supply = <&vcc_io>;
 	};
 
+	vcc_flash: flash-regulator {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc_flash";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		vin-supply = <&vcc_io>;
+	};
+
+	vdd_log: vdd_log {
+		compatible = "regulator-fixed";
+		regulator-min-microvolt = <1000000>;
+		regulator-max-microvolt = <1000000>;
+		regulator-name = "vdd_logic";
+	};
+	
+	rk3288-gpiomem {
+		compatible = "rockchip,rk3288-gpiomem";
+		reg = <0x0 0xff750000 0x0 0x1000>;
+		status = "okay";
+	};
+	
 	vcc_cam: vcc-camera {
 		compatible = "regulator-fixed";
 		regulator-name = "vcc_cam";
@@ -192,6 +216,11 @@
 	};
 };
 
+&cif_isp0 {
+	rockchip,camera-modules-attached = <&camera0 &camera1>;
+	status = "okay";
+};
+
 &cpu0 {
 	cpu0-supply = <&vdd_cpu>;
 };
@@ -447,6 +476,10 @@
 	};
 };
 
+&i2c1 {
+	status = "okay";
+};
+
 &i2c2 {
 	status = "okay";
 
@@ -538,6 +571,9 @@
 &spi2 {
 	status = "okay";
 	max-freq = <50000000>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi2_clk &spi2_tx &spi2_rx &spi2_cs0 &spi2_cs1>;
+	#address-cells = <2>;
 
 	spidev@0 {
 		compatible = "rockchip,spi_tinker";
@@ -561,7 +597,9 @@
 
 &io_domains {
 	status = "okay";
-
+	rockchip,grf = <&grf>;
+	flash0-supply = <&vcc_flash>;
+	gpio30-supply = <&vcc_io>;
 	sdcard-supply = <&vccio_sd>;
 	wifi-supply = <&vcc_18>;
 };
@@ -612,6 +650,20 @@
 	vqmmc-supply = <&vccio_sd>;
 };
 
+&emmc {
+	bus-width = <8>;
+	cap-mmc-highspeed;
+	disable-wp;
+	non-removable;
+	num-slots = <1>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&emmc_clk &emmc_cmd &emmc_pwr &emmc_bus8>;
+	max-frequency = <150000000>;
+	mmc-hs200-1_8v;
+	mmc-ddr-1_8v;
+	status = "okay";
+};
+
 &tsadc {
 	rockchip,hw-tshut-mode = <1>; /* tshut mode 0:CRU 1:GPIO */
 	rockchip,hw-tshut-polarity = <1>; /* tshut polarity 0:LOW 1:HIGH */
