diff --git a/arch/arm/boot/dts/rk3288-miniarm.dts b/arch/arm/boot/dts/rk3288-miniarm.dts
index 24a4827c..89fc18c1 100644
--- a/arch/arm/boot/dts/rk3288-miniarm.dts
+++ b/arch/arm/boot/dts/rk3288-miniarm.dts
@@ -498,17 +498,23 @@
 		reg = <0x36>;
 		clocks = <&ext_cam_clk>;
 		status = "okay";
+
+		port {
+			ov5647_out: endpoint {
+				remote-endpoint = <&mipi_dphy_in>;
+				data-lanes = <1 2>;
+			};
+		};
 	};
 
 	camera1: imx219@10 {
 		compatible = "sony,imx219";
 		reg = <0x10>;
 		clocks = <&ext_cam_clk>;
-		status = "okay";
-
+		status = "disabled";
 		port {
 			imx219_out: endpoint {
-				remote-endpoint = <&imx219_in>;
+				remote-endpoint;
 				data-lanes = <1 2>;
 			};
 		};
@@ -562,8 +568,8 @@
 		port@0 {
 			reg = <0>;
 
-			imx219_in: endpoint {
-				remote-endpoint = <&imx219_out>;
+			mipi_dphy_in: endpoint {
+				remote-endpoint = <&ov5647_out>;
 				data-lanes = <1 2>;
 			};
 		};
