diff --git a/arch/arm/boot/dts/overlay/Makefile b/arch/arm/boot/dts/overlay/Makefile
index 56d8cb18..c46f29d6 100644
--- a/arch/arm/boot/dts/overlay/Makefile
+++ b/arch/arm/boot/dts/overlay/Makefile
@@ -11,7 +11,8 @@ dtbo-$(CONFIG_ARCH_ROCKCHIP) += \
 	rockchip-uart2.dtbo	\
 	rockchip-uart3.dtbo	\
 	rockchip-uart4.dtbo	\
-	rockchip-w1-gpio.dtbo
+	rockchip-w1-gpio.dtbo   \
+	rockchip-imx219.dtbo
 
 scr-$(CONFIG_ARCH_ROCKCHIP) += \
        rockchip-fixup.scr
diff --git a/arch/arm/boot/dts/overlay/README.rockchip-overlays b/arch/arm/boot/dts/overlay/README.rockchip-overlays
index df4559ca..4246b8b8 100644
--- a/arch/arm/boot/dts/overlay/README.rockchip-overlays
+++ b/arch/arm/boot/dts/overlay/README.rockchip-overlays
@@ -20,6 +20,7 @@ rockchip (Rockchip)
 - uart3
 - uart4
 - w1-gpio
+- imx-219
 
 ### Overlay details:
 
@@ -75,4 +76,8 @@ Conflicts with spi0
 
 Activates 1-wire gpio master on GPIO0 17
 
+### imx219
+
+Sets imx219 (RPi cam V2.1) as default camera on CSI 
+
 
diff --git a/arch/arm/boot/dts/overlay/rockchip-imx219.dts b/arch/arm/boot/dts/overlay/rockchip-imx219.dts
new file mode 100644
index 00000000..1fc6af4d
--- /dev/null
+++ b/arch/arm/boot/dts/overlay/rockchip-imx219.dts
@@ -0,0 +1,57 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	compatible = "rockchip,rk3288-miniarm", "rockchip,rk3288";
+
+	fragment@0 {
+		target = <&i2c2>;
+		__overlay__ {
+
+			camera0: ov5647@36 {
+				compatible = "ovti,ov5647";
+				reg = <0x36>;
+				clocks = <&ext_cam_clk>;
+				status = "disabled";
+
+				port {
+					ov5647_out: endpoint {
+					};
+				};
+			};
+
+			camera1: imx219@10 {
+				compatible = "sony,imx219";
+				reg = <0x10>;
+				clocks = <&ext_cam_clk>;
+				status = "okay";
+				port {
+					imx219_out: endpoint {
+						remote-endpoint = <&mipi_dphy_in>;
+						data-lanes = <1 2>;
+					};
+				};
+			};
+
+		};
+	};
+
+	fragment@1 {
+		target = <&mipi_phy_rx0>;
+		__overlay__ {
+			ports {
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				port@0 {
+					reg = <0>;
+
+					mipi_dphy_in: endpoint {
+						remote-endpoint = <&imx219_out>;
+						data-lanes = <1 2>;
+					};
+				};
+			};
+		};
+	};
+};
