diff --git a/arch/arm/boot/dts/rk3288-miniarm.dts b/arch/arm/boot/dts/rk3288-miniarm.dts
index de6c4a6..30f0c46 100644
--- a/arch/arm/boot/dts/rk3288-miniarm.dts
+++ b/arch/arm/boot/dts/rk3288-miniarm.dts
@@ -42,7 +42,6 @@
 #include <dt-bindings/clock/rockchip,rk808.h>
 #include "rk3288.dtsi"
 #include "rk3288-rkisp1.dtsi"
-#include "rk3288-linux.dtsi"
 
 / {
 	compatible = "rockchip,rk3288-miniarm", "rockchip,rk3288";
