From d2a852603b364b90903dc0a9040d4b11f79b8fea Mon Sep 17 00:00:00 2001
From: The-going <48602507+The-going@users.noreply.github.com>
Date: Thu, 6 Feb 2025 15:11:40 +0300
Subject: [PATCH 6/6] h616: GPU enable. Magic register.

---
 arch/arm/mach-sunxi/clock_sun50i_h6.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/arch/arm/mach-sunxi/clock_sun50i_h6.c b/arch/arm/mach-sunxi/clock_sun50i_h6.c
index b424a7893ea..9ed5852d547 100644
--- a/arch/arm/mach-sunxi/clock_sun50i_h6.c
+++ b/arch/arm/mach-sunxi/clock_sun50i_h6.c
@@ -15,6 +15,8 @@ void clock_init_safe(void)
 		/* this seems to enable PLLs on H616 */
 		setbits_le32(&prcm->sys_pwroff_gating, 0x10);
 		setbits_le32(&prcm->res_cal_ctrl, 2);
+		/* enable GPU */
+		writel(0, 0x7010254);
 	}
 
 	if (IS_ENABLED(CONFIG_MACH_SUN50I_H616) ||
-- 
2.35.3

